{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "Transcriptome Tutorial.ipynb",
      "version": "0.3.2",
      "provenance": []
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.7.1"
    },
    "kernelspec": {
      "display_name": "Python 3",
      "language": "python",
      "name": "python3"
    }
  },
  "cells": [
    {
      "metadata": {
        "colab_type": "text",
        "id": "ROl7-Rhd9if9"
      },
      "cell_type": "markdown",
      "source": [
        "# Transcriptomics Tutorial\n",
        "\n",
        "\n",
        "## Introduction\n",
        "> Cell phenotype is controlled by gene expression regulation, a phenomenon discovered by Barbara McClintock in early 1950s, later described by  Francois Jacob and Jacques Monod in bacteria. Since the discovery of the entire human DNA sequence, scientists have now the knowledge and technology to quantify precisely gene expression at whole genome scale. A particularly interesting layer of regulation occurs at the transcriptome level. The transcriptome is the set of all RNA molecules in the cell. Cellular phenotype is particularly impacted by the pool of messenger RNAs (mRNAs), which results from the transcription of specific genes and serves as template for protein translation. Through the use of transcriptomics techniques, we can determine the abundance of each mRNA in a cell, therefore determining the levels of gene expression. Quantifying gene expression levels in our cells helps us understand important biological phenomena like cell differentiation and mechanisms of human disease.\n",
        "\n",
        "> There are two main transcriptomics technologies use to measure gene expression: (1) [Microarrays](https://en.wikipedia.org/wiki/DNA_microarray) is an older technique used to measure relative abundance of gene expression. We will not cover microarrays here, but this [video](https://www.youtube.com/watch?v=6ZzFihESjp0) does a great job explaining how microarrays works, and (2) RNA sequencing (RNA-seq). In this tutorial, we will focus only focus on RNA-seq data analysis.\n",
        "\n",
        "\n",
        "##  RNA-seq\n",
        "\n",
        "> [RNA-seq](https://en.wikipedia.org/wiki/RNA-Seq) is a technology to quantify gene expression by determining the abundance of specific RNA sequences. There are three main steps to RNA-seq: library preparation, sequencing, and data analysis. \n",
        "\n",
        "> Let's start with the steps happening in the lab. First, we isolate mRNA. Then, using the enzyme reverse transcriptase, mRNA is converted into its complementary DNA strand, cDNA. The cDNA is then cut up into small fragments, usually a few hundred base-pairs long. [Next generation sequencing technologies](https://en.wikipedia.org/wiki/Illumina_dye_sequencing) use fluorescence-tagged nucleotides that bind to short cDNA sequences to recapitulate the sequence of the sample by sequencing, in a massively parallel manner. Each sequenced read (call *reads*) are stored into a file of [FASTQ format](https://en.wikipedia.org/wiki/FASTQ_format).\n",
        "\n",
        "> Later, as part of the data analysis pipeline, theses files containing the sequence of the reads are then cleaned using quality control pipelines, then mapped their sequence to a specific genome / transcriptome / set of features.\n",
        "\n",
        "## File Structure\n",
        "\n",
        "> Organization is a very important skill for any bioinformatics project that is oftentimes overlooked. There is no standard file hierarchy to use, but we will use a file structure that is fairly common among many research scientists:\n",
        "\n",
        "<html>\n",
        "<div style=\"width: 50%;margin: 0 auto; background-color: rgb(242, 242, 242);\">\n",
        "    <code><i>project_name</i>/\n",
        "    data/\n",
        "        BAM/\n",
        "        cleanFASTQ/\n",
        "        cufflinks/\n",
        "        FASTQ/\n",
        "        kallisto/\n",
        "        microbesOnline/\n",
        "        seqData/\n",
        "        transcriptomeIndex/</code>\n",
        "</div>\n",
        "</html>\n",
        "\n",
        "> Also, we will store all downloaded software in a  folder called *software* in the home directory, ~/software, and add binaries of downloaded software to the [PATH](https://en.wikipedia.org/wiki/PATH_(variable)). Instructions for adding to your PATH on Mac can be found [here](https://coolestguidesontheplanet.com/add-shell-path-osx/).\n",
        "\n",
        "## Setup\n",
        "\n",
        "> Below, we will assign the path to the working directory as a global variable that will be used throughout the rest of the notebook."
      ]
    },
    {
      "metadata": {
        "id": "7Ll4PktyDRmv",
        "colab_type": "code",
        "colab": {}
      },
      "cell_type": "code",
      "source": [
        "###IMPORTANT: if you ever restart this notebook, you MUST rerun this code cell\n",
        "import os\n",
        "\n",
        "#feel free to change this location, because we will be working in this folder for the entire tutorial\n",
        "os.environ['WORKDIR'] = './data' "
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "metadata": {
        "id": "vNo35L2jDRmy",
        "colab_type": "code",
        "colab": {}
      },
      "cell_type": "code",
      "source": [
        "#create the directory if doesn't exist already\n",
        "!mkdir -p $WORKDIR"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "metadata": {
        "colab_type": "text",
        "id": "kdsef3qW9if-"
      },
      "cell_type": "markdown",
      "source": [
        "\n",
        "# RNA-Seq Data Analysis\n",
        "> In this tutorial, we will use a subset of samples from a published work, where they investigated how microbes adapt to new environments ([PMID 28854640](https://www.ncbi.nlm.nih.gov/pubmed/28854640)). Following standard practices in science, the authors made all data required to reproduce their findings, including the reads for several transcriptome samples, publically available (SRA Project [SRP106486](https://trace.ncbi.nlm.nih.gov/Traces/sra/?study=SRP106486)).\n",
        "\n",
        "First, we need to download the raw data, which in this case is a file of 500 MB. For a connection speed of around 6 MB/s, the file will download in about 2 minutes, but download time will vary depending on your internet connection. Afterwards, move the file into the $WORKDIR/seqData/ folder."
      ]
    },
    {
      "metadata": {
        "colab_type": "code",
        "id": "Rb1gIaCw9if-",
        "outputId": "670d955f-14dd-452a-f397-3c16de0c90ea",
        "scrolled": true,
        "colab": {}
      },
      "cell_type": "code",
      "source": [
        "!wget ftp://ftp-trace.ncbi.nih.gov/sra/sra-instant/reads/ByRun/sra/SRR/SRR551/SRR5511057/SRR5511057.sra"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "--2019-02-05 17:30:41--  ftp://ftp-trace.ncbi.nih.gov/sra/sra-instant/reads/ByRun/sra/SRR/SRR551/SRR5511057/SRR5511057.sra\n",
            "           => ‘SRR5511057.sra’\n",
            "Resolving ftp-trace.ncbi.nih.gov (ftp-trace.ncbi.nih.gov)... 130.14.250.7\n",
            "Connecting to ftp-trace.ncbi.nih.gov (ftp-trace.ncbi.nih.gov)|130.14.250.7|:21... connected.\n",
            "Logging in as anonymous ... Logged in!\n",
            "==> SYST ... done.    ==> PWD ... done.\n",
            "==> TYPE I ... done.  ==> CWD (1) /sra/sra-instant/reads/ByRun/sra/SRR/SRR551/SRR5511057 ... done.\n",
            "==> SIZE SRR5511057.sra ... 455056642\n",
            "==> PASV ... done.    ==> RETR SRR5511057.sra ... done.\n",
            "Length: 455056642 (434M) (unauthoritative)\n",
            "\n",
            "SRR5511057.sra      100%[===================>] 433.98M  6.28MB/s    in 83s     \n",
            "\n",
            "2019-02-05 17:32:05 (5.24 MB/s) - ‘SRR5511057.sra’ saved [455056642]\n",
            "\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "metadata": {
        "colab_type": "text",
        "id": "UreiWlGA9igC"
      },
      "cell_type": "markdown",
      "source": [
        "***\n",
        "If the above command returned `wget: command not found`, means you don't have the command `wget` in your machine. Then, try installing Homebrew, run `brew install wget` in terminal, and then try the above command again."
      ]
    },
    {
      "metadata": {
        "colab_type": "text",
        "id": "RwOtgqis9igC"
      },
      "cell_type": "markdown",
      "source": [
        "## Preparing the file\n",
        "You may notice that we downloaded a `.sra` file. This file format contains the raw sequencing data and alignment information in a compressed form. We will convert this into [FASTQ](https://en.wikipedia.org/wiki/FASTQ_format) format, which is the standard file format for storing RNA-seq raw data. FASTQ is a text-based format for storing the biological sequence and its corresponding quality scores. Each read is encoded with 4 lines in strict order:\n",
        "\n",
        "* Line 1: beings with '@' character with an identifier for the sequence and an optional description.\n",
        "* Line 2: raw nucleotide sequence.\n",
        "* Line 3: begins with '+' character with the same sequence identifier and additional optional description.\n",
        "* Line 4: quality of each nucleotide.\n",
        "\n",
        "To help us convert this file into FASTQ, you will need to download [SRA Toolkit](https://www.ncbi.nlm.nih.gov/books/NBK158900/).\n",
        "Although this toolkit contains many helpful tools to help us work with sequencing data, there is only one command we need to learn for now:\n",
        "\n",
        "`fastq-dump --split-files -O <outputDir> <sraFile>`\n",
        "\n",
        "This command converts the .sra file into FASTQ format. The --split-files tag specifies whether our data was obtained using single-end sequencing or paired-end sequencing. Single-end sequencing means that each hybridized fragment is read once, while paired-end sequencing means that each hybridized fragment is read twice (once from each end). Paired-end sequencing results in twice as many reads compared to single-end sequencing, so is therefore more costly, but improves accuracy and the quality of the sequencing."
      ]
    },
    {
      "metadata": {
        "colab_type": "code",
        "id": "JuE8eZgm9igD",
        "outputId": "d806e73a-e5c5-4e3f-d50a-8243269e325b",
        "colab": {}
      },
      "cell_type": "code",
      "source": [
        "!fastq-dump --split-files -O $WORKDIR/FASTQ $WORKDIR/seqData/SRR5511057.sra"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Read 7370822 spots for ./data/seqData/SRR5511057.sra\r\n",
            "Written 7370822 spots for ./data/seqData/SRR5511057.sra\r\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "metadata": {
        "colab_type": "text",
        "id": "0VRA2GMM9igF"
      },
      "cell_type": "markdown",
      "source": [
        "Our data was sequenced using paired-end sequencing, so we used the `--split-files` flag. You should see two new files in your $WORKDIR/FASTQ folder: `SRR5511057_1.fastq` and `SRR5511057_2.fastq`. If you inspect the SRR5511057_1.fastq file with the `head` command, we can see what our reads look like in FASTQ format. Here is the first read, corresponding to the first 4 lines of the file (notice that line 2 and line 4 are exactly the same length! Why does that make sense?<sub>(1)</sub>):\n",
        "\n",
        "<div style=\"border:solid 1px; padding:5px\">\n",
        "@SRR5511057.1 1 length=76 <br>\n",
        "GTCCACACAAGCCAAATGGTGAAATGCAGTTCTGGGTGTGCCTAGGTATAAGGATTTGTCCTTCTTGCTGACAGCT <br>\n",
        "+SRR5511057.1 1 length=76 <br>\n",
        "AAAAAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE/EEEEEEEEEEEEEAEEEEE\n",
        "</div>"
      ]
    },
    {
      "metadata": {
        "colab_type": "text",
        "id": "8E9OqEqf9igG"
      },
      "cell_type": "markdown",
      "source": [
        "## Converting Raw Data into Usable Data\n",
        "\n",
        "There are three overall steps in RNA-seq transcriptome qualification:\n",
        "1. **Cleaning reads.** Reads typically have a length of 75-250 bp (base-pairs). Due to molecular noise, not all positions of the read have the same likelihood of that event being true. For this reason, we have to clean and trim the data.\n",
        "2. **Mapping reads.** We need to associate reads to our reference features, e.g., transcripts. For this task, we need the FASTQ files and a reference annotation: (1) a set of files that define the genome, typically FASTA files and (2) a file defining where transcripts are located in the genome, typically a GFF3 file. The output of the mapping algorithms is a BAM file, which is a compressed binary file that contains the aligned reads to the reference.\n",
        "3. **Quantifying reads.** Next, we need to associate an expression value to each feature, i.e., quantify the level of expression of our transcripts from the reads alignment.\n",
        "\n",
        "There are 3 major units for quantifying reads: \n",
        "- FPKM: is useful for visualization across genes and samples because it accounts for the variability of transcript length and experimental sequencing depth. This is useful, since longer transcripts will produce more transcript fragments by chance, and in turn will have more reads associated with it since the longer the transcript.\n",
        "- TPM: is similar to FPKM, but it normalizes each gene expression to a constant derived from the sum of expression across all genes. This is helpful for comparison among samples where effective transcript lenght changes. Here's a very helpful [blog post and video](https://www.rna-seqblog.com/rpkm-fpkm-and-tpm-clearly-explained/) understanding the difference between TPM and FPKM.\n",
        "- _counts_: is the raw number of reads that match to the transcript. It is useful for statistical purposes. \n"
      ]
    },
    {
      "metadata": {
        "colab_type": "text",
        "id": "O1GxYpkJ9igH"
      },
      "cell_type": "markdown",
      "source": [
        "***\n",
        "In the following section, we will walk through 3 different pipelines that we can use to clean, map, and quantify the reads."
      ]
    },
    {
      "metadata": {
        "colab_type": "text",
        "id": "Yi0HaACA9igH"
      },
      "cell_type": "markdown",
      "source": [
        "### Pipeline A: Trimmomatic/STAR/cufflinks\n",
        "\n",
        "**1a.** [Trimmomatic](http://www.usadellab.org/cms/?page=trimmomatic) is a Java software to trim and remove low quality reads. It also removes adapter sequences from library preparation. After downloading the software from its site, extract the zip file and move it into the ~/software folder. Next, execute the following command:"
      ]
    },
    {
      "metadata": {
        "colab_type": "code",
        "id": "qnmlB-Bx9igK",
        "outputId": "551c2545-62dd-4f73-9318-9f4e731500f7",
        "colab": {}
      },
      "cell_type": "code",
      "source": [
        "!java -jar ~/software/Trimmomatic-0.38/trimmomatic-0.38.jar PE -phred33 \\\n",
        "    $WORKDIR/FASTQ/SRR5511057_1.fastq $WORKDIR/FASTQ/SRR5511057_2.fastq \\\n",
        "    $WORKDIR/cleanFASTQ/SRR5511057.forward_paired.fastq $WORKDIR/cleanFASTQ/SRR5511057.forward_unpaired.fastq \\\n",
        "    $WORKDIR/cleanFASTQ/SRR5511057.reverse_paired.fastq $WORKDIR/cleanFASTQ/SRR5511057.reverse_unpaired.fastq \\\n",
        "    LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "TrimmomaticPE: Started with arguments:\n",
            " -phred33 ./data/sequence_data/SRR5511057_1.fastq ./data/sequence_data/SRR5511057_2.fastq ./data/sequence_data/SRR5511057.forward_paired.fastq ./data/sequence_data/SRR5511057.forward_unpaired.fastq ./data/sequence_data/SRR5511057.reverse_paired.fastq ./data/sequence_data/SRR5511057.reverse_unpaired.fastq LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36\n",
            "Multiple cores found: Using 4 threads\n",
            "Input Read Pairs: 7370822 Both Surviving: 6915652 (93.82%) Forward Only Surviving: 369084 (5.01%) Reverse Only Surviving: 56673 (0.77%) Dropped: 29413 (0.40%)\n",
            "TrimmomaticPE: Completed successfully\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "metadata": {
        "colab_type": "text",
        "id": "35Jmyzga9igN"
      },
      "cell_type": "markdown",
      "source": [
        "Let's look over the arguments used in this command. \n",
        "- `PE` stands for paired end, which is the type of sequencing our sample data came from. \n",
        "- `-phred33` [Phred](https://en.wikipedia.org/wiki/Phred_quality_score) + 33 Quality Score.\n",
        "- `$WORKDIR/FASTQ/SRR5511057_1.fastq` `$WORKDIR/FASTQ/SRR5511057_2.fastq` is the relative location for our input data.\n",
        "- `$WORKDIR/cleanFASTQ/SRR5511057.forward_paired.fastq` `$WORKDIR/cleanFASTQ/SRR5511057.forward_unpaired.fastq` is the relative location for our output forward paired and unpaired data, respectively.\n",
        "- `$WORKDIR/cleanFASTQ/SRR5511057.reverse_paired.fastq` `$WORKDIR/cleanFASTQ/SRR5511057.reverse_unpaired.fastq` is the relative location for our output reverse paired and unpaired data, respectively.\n",
        "- `LEADING:3` removes leading bases with a low quality (below quality 3).\n",
        "- `TRAILING:3` removes trailing bases with a low quality (below quality 3).\n",
        "- `SLIDINGWINDOW:4:15` scans the read with a 4-base sliding window, cutting when the average quality per base drops below 15.\n",
        "- `MINLEN:36` drops reads below the 36 bases long.\n",
        "\n",
        "DNA is double-stranded, but one strand is designated as the \"forward strand\" while the other strand is designated as the \"reverse strand\". Another common term for this is plus/minus strands. Since paired-end sequencing reads match both directions, the reads that are complementary to the forward strand are called \"reverse reads\", while the reads that are complementary with the reverse strand are called \"forward reads\". \n",
        "\n",
        "In paired-end sequencing, reads come in pairs (each fragment yields two reads, one from each end), so if one of the reads is trimmed off by the criteria we set when running Trimmomatic, Trimmomatic will automatically remove the corresponding read and label both reads as \"unpaired\".\n",
        "***"
      ]
    },
    {
      "metadata": {
        "colab_type": "text",
        "id": "8Q4YJ-Le9igN"
      },
      "cell_type": "markdown",
      "source": [
        "**1b.** [Star](https://github.com/alexdobin/STAR) (Spliced Transcripts Alignment to a Reference) is a C++ program to align reads to a reference genome. For performance efficiency, its first step is building an index of our reference genome (this step only needs to be done once!).\n",
        "\n",
        "First, we will need to download STAR software and untar the file. Afterwards, remember to place STAR in the ~/software folder and add it to your PATH."
      ]
    },
    {
      "metadata": {
        "colab_type": "code",
        "id": "kXU2Pe5D9igO",
        "outputId": "25522bd5-1a98-4221-c418-f68f851ee898",
        "colab": {}
      },
      "cell_type": "code",
      "source": [
        "!wget https://github.com/alexdobin/STAR/archive/2.7.0d.tar.gz\n",
        "!tar -xzf 2.7.0d.tar.gz"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "--2019-02-19 10:27:34--  https://github.com/alexdobin/STAR/archive/2.7.0d.tar.gz\n",
            "Resolving github.com (github.com)... 192.30.255.113, 192.30.255.112\n",
            "Connecting to github.com (github.com)|192.30.255.113|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://codeload.github.com/alexdobin/STAR/tar.gz/2.7.0d [following]\n",
            "--2019-02-19 10:27:35--  https://codeload.github.com/alexdobin/STAR/tar.gz/2.7.0d\n",
            "Resolving codeload.github.com (codeload.github.com)... 192.30.255.121, 192.30.255.120\n",
            "Connecting to codeload.github.com (codeload.github.com)|192.30.255.121|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: unspecified [application/x-gzip]\n",
            "Saving to: ‘2.7.0d.tar.gz’\n",
            "\n",
            "2.7.0d.tar.gz           [        <=>         ]   7.69M  4.94MB/s    in 1.6s    \n",
            "\n",
            "2019-02-19 10:27:36 (4.94 MB/s) - ‘2.7.0d.tar.gz’ saved [8062298]\n",
            "\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "metadata": {
        "colab_type": "text",
        "id": "Jj_egHZu9igQ"
      },
      "cell_type": "markdown",
      "source": [
        "**1c.** The first step to mapping our reads is to generate an index. Since we are interested in the expression of transcripts, we will download the reference transcriptome sequence from [MicrobesOnline](http://www.microbesonline.org/), an online database of microbial genomic information. Our data was generated in a sample from yeast, or scientifically known as *Saccharomyces ceravisiae*, so we need the FASTA transcriptome for yeast from [here](http://www.microbesonline.org/cgi-bin/genomeInfo.cgi?tId=4932). Unzip the file, and move the resulting file to the $WORKDIR/microbesOnline folder. Then, we will need to make a new folder to store the transcriptome indices."
      ]
    },
    {
      "metadata": {
        "colab_type": "code",
        "id": "AinjBoW99igQ",
        "colab": {}
      },
      "cell_type": "code",
      "source": [
        "!mkdir $WORKDIR/transcriptomeIndex"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "metadata": {
        "colab_type": "text",
        "id": "z8X1ecOw9igS"
      },
      "cell_type": "markdown",
      "source": [
        "To build the index, we will run the following command."
      ]
    },
    {
      "metadata": {
        "colab_type": "code",
        "id": "5ntrzPe39igS",
        "outputId": "cc2c9d2c-067f-4c44-b15b-b39ee2962f01",
        "colab": {}
      },
      "cell_type": "code",
      "source": [
        "!STAR --runThreadN 8 --runMode genomeGenerate \\\n",
        "    --genomeDir $WORKDIR/transcriptomeIndex --genomeFastaFiles $WORKDIR/microbesOnline/4932.transcriptomes.fasta"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Feb 19 15:13:26 ..... started STAR run\n",
            "Feb 19 15:13:26 ... starting to generate Genome files\n",
            "Feb 19 15:13:35 ... starting to sort Suffix Array. This may take a long time...\n",
            "Feb 19 15:13:40 ... sorting Suffix Array chunks and saving them to disk...\n",
            "Feb 19 15:13:46 ... loading chunks from disk, packing SA...\n",
            "Feb 19 15:13:48 ... finished generating suffix array\n",
            "Feb 19 15:13:48 ... generating Suffix Array index\n",
            "Feb 19 15:13:55 ... completed Suffix Array index\n",
            "Feb 19 15:13:55 ... writing Genome to disk ...\n",
            "Feb 19 15:13:57 ... writing Suffix Array to disk ...\n",
            "Feb 19 15:13:57 ... writing SAindex to disk\n",
            "Feb 19 15:13:59 ..... finished successfully\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "metadata": {
        "colab_type": "text",
        "id": "hSvfdu0g9igU"
      },
      "cell_type": "markdown",
      "source": [
        "Each -- defines a new parameter, which are options we specify to STAR\n",
        "- `--runThreadN 8` specifies the number of threads on your computer (hint: run `sysctl -n hw.ncpu` to find out!).\n",
        "- `--runMode genomeGenerate` means that we are building the genome index.\n",
        "- `--genomeDir $WORKDIR/transcriptomeIndex` saves the output the the folder we just created.\n",
        "- `--genomeFastaFiles` are the reference genome sequences (in this case we are using the transcriptome) we are building our index from.\n",
        "\n",
        "**1d.** Now, we map our reads to the transcriptome!"
      ]
    },
    {
      "metadata": {
        "colab_type": "code",
        "id": "S-UEMAfA9igV",
        "outputId": "be1b579c-b488-479d-c164-5aa5d3a95b81",
        "scrolled": true,
        "colab": {}
      },
      "cell_type": "code",
      "source": [
        "!STAR --runThreadN 4 --genomeDir $WORKDIR/transcriptomeIndex \\\n",
        "    --readFilesIn $WORKDIR/cleanFASTQ/SRR5511057.forward_paired.fastq $WORKDIR/cleanFASTQ/SRR5511057.reverse_paired.fastq \\\n",
        "     --outSAMtype BAM SortedByCoordinate --outFileNamePrefix $WORKDIR/BAM/\n"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Apr 11 22:45:32 ..... started STAR run\n",
            "Apr 11 22:45:32 ..... loading genome\n",
            "Apr 11 22:45:38 ..... started mapping\n",
            "Apr 11 22:46:48 ..... finished mapping\n",
            "Apr 11 22:46:48 ..... started sorting BAM\n",
            "Apr 11 22:46:56 ..... finished successfully\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "metadata": {
        "colab_type": "text",
        "id": "0IqjHKh19igX"
      },
      "cell_type": "markdown",
      "source": [
        "If you ran into this error: \n",
        "\n",
        "`\n",
        "BAMoutput.cpp:27:BAMoutput: exiting because of *OUTPUT FILE* error: could not create output file ./data/BAM_STARtmp//BAMsort/4/48`\n",
        "\n",
        "SOLUTION: check that the path exists and you have write permission for this file. Also check `ulimit -n` and increase it to allow more open files. Also, it may be worth lowering the number of threads or running `ulimit -n 10000` in the terminal.\n",
        "***"
      ]
    },
    {
      "metadata": {
        "colab_type": "text",
        "id": "dPaDqe8A9igY"
      },
      "cell_type": "markdown",
      "source": [
        "Let's look at the options used in this command:\n",
        "- `--runThreadN 4` specifies the number of threads you want to use (run `sysctl -n hw.ncpu` to find the number of threads on your computer!).\n",
        "- `--genomeDir $WORKDIR/transcriptomeIndex` directory where genome index was generated.\n",
        "- `--readFilesIn` input sequence data to be mapped.\n",
        "- `--outSAMtype BAM SortedByCoordinate` outputs a BAM file with sorted aligned reads.\n",
        "- `--outFileNamePrefix` prefix to add to the output file.\n",
        "\n",
        "Our aligned reads are stored in a BAM format, which is a compressed binary version of [SAM](https://en.wikipedia.org/wiki/SAM_(file_format)), a text-based format for storing aligned reads.\n",
        "\n",
        "**1e.** We will now use [cufflinks](http://cole-trapnell-lab.github.io/cufflinks/) to quantify the reads per feature. Download the latest version [here](http://cole-trapnell-lab.github.io/cufflinks/install/), and add the directory to your Path. "
      ]
    },
    {
      "metadata": {
        "colab_type": "code",
        "id": "SnzCISia9igY",
        "outputId": "675af77d-d1b8-461b-bd40-07ec8cf95437",
        "colab": {}
      },
      "cell_type": "code",
      "source": [
        "!cufflinks $WORKDIR/BAM/Aligned.sortedByCoord.out.bam --library-type fr-firststrand"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Warning: Could not connect to update server to verify current version. Please check at the Cufflinks website (http://cufflinks.cbcb.umd.edu).\n",
            "[22:47:40] Inspecting reads and determining fragment length distribution.\n",
            "> Processed 6316 loci.                         [*************************] 100%\n",
            "> Map Properties:\n",
            ">\tNormalized Map Mass: 6039610.00\n",
            ">\tRaw Map Mass: 6039610.00\n",
            ">\tFragment Length Distribution: Empirical (learned)\n",
            ">\t              Estimated Mean: 157.12\n",
            ">\t           Estimated Std Dev: 62.28\n",
            "[22:48:28] Assembling transcripts and estimating abundances.\n",
            "> Processed 6316 loci.                         [*************************] 100%\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "metadata": {
        "id": "iVbjO5pbDRnL",
        "colab_type": "text"
      },
      "cell_type": "markdown",
      "source": [
        "- `$WORKDIR/BAM/Aligned.sortedByCoord.out.bam` is the file of RNA-seq read alignments in BAM or SAM format\n",
        "- `--library-type fr-firststrand` specifies the library used to prepare our data, we used Illumina's stranded Tru-Seq which is fr-firststrand"
      ]
    },
    {
      "metadata": {
        "id": "sPV06Z-mDRnM",
        "colab_type": "text"
      },
      "cell_type": "markdown",
      "source": [
        "Cufflinks should generate 4 files: genes.fpkm_tracking, isoforms.fpkm_tracking, skipped.gtf, and transcripts.gtf. Move these files into your $WORKDIR/cufflinks folder. \n",
        "\n",
        "We are interested in the genes.fpkm_tracking file. In our example, we used the FASTA transcriptome to build the STAR index, so the genes.fpkm_tracking file quantifies the expression of every transcript. However, if we were to use the FASTA genome and a GFF3 annotation file to build the STAR index, genes.fpkm_tracking would quantify gene expression while isoforms.fpkm_tracking would quantify the expression for all possible exons.\n",
        "\n",
        "You can open genes.fpkm_tracking using any text editor. You should have a table that looks like the following:\n",
        "\n",
        "|tracking_id\t| class_code | nearest_ref_id | gene_id | gene_short_name | tss_id | locus | length | coverage | FPKM | FPKM_conf_lo | FPKM_conf_hi | FPKM_status |\n",
        "| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |\n",
        "| CUFF.1 | - | - | CUFF.1 | - | - | VIMSS6580636:30-1540 | - | - | 3.54675 | 2.207 | 4.8865 | OK |\n",
        "| CUFF.2 | - | - | CUFF.2 | - | - | VIMSS6580642:3238-3969 | - | - | 19.8471 | 14.9253 | 24.769 | OK |\n",
        "| ... |\n",
        "\n",
        "In this tutorial, we only care about a few of the columns:\n",
        "1. `gene_id` is the Cufflinks gene id\n",
        "2. `locus` is where the reads was aligned to on the transcriptome, for example: VIMSS6580636:30-1540 means on VIMSS6580636 transcript from the 30th to the 1540th base pair\n",
        "3. `FPKM` is level of transcript expression in FPKM\n",
        "4. `FPKM_conf_lo` is the lower bound of the 95% confidence interval for FPKM\n",
        "5. `FPKM_conf_hi` is the upper bound of the 95% confidence interval for FPKM\n",
        "6. `FPKM_status` is fairly self-explanatory\n",
        "\n",
        "By looking at the FPKM value, we can gain a general understanding of the level of expression of each transcript:\n",
        "- 0-10 is considered low\n",
        "- 10-100 is considered medium\n",
        "- 100-1000 is considered high\n",
        "- 1000+ is considered very high\n",
        "\n",
        "**Question:** If we used pipeline A to measure gene expression in humans and built the STAR index using the entire human genome and an annotation file, would the genes.fpkm_tracking file or isoforms.fpkm_tracking file have more entries?<sub>(2)</sub>"
      ]
    },
    {
      "metadata": {
        "colab_type": "text",
        "id": "Ftqz2Ex_9igb"
      },
      "cell_type": "markdown",
      "source": [
        "### Pipeline B: Trimmomatic/killisto\n",
        "If you have walked through pipeline A, you can skip Trimmomatic because you have already cleaned the reads.\n",
        "If you still need to clean your data, see instructions for [Trimmomatic](http://www.usadellab.org/cms/?page=trimmomatic) in pipeline A, then return here.\n",
        "\n",
        "[Kallisto](https://pachterlab.github.io/kallisto/) is a program that appeared in [2015](https://liorpachter.wordpress.com/2015/05/10/near-optimal-rna-seq-quantification-with-kallisto/) developed in Lior Patcher's group. It is based on _pseudo-alignment_ using [k-mers](https://en.wikipedia.org/wiki/K-mer) instead of reads. It is fast, accurate, and robust. And simple to use. All you need it to create is an index of your transcriptome (or any target reference sequence) and you will be ready to apply it on your FASTQ files. This [video](https://www.youtube.com/watch?v=94wphB3GKBM) has a great explanation of how kallisto differs from other mapping tools.\n",
        "\n",
        "**2a.** The first step is to create a kallisto index, and we can use the same transcriptome FASTA file from pipeline A."
      ]
    },
    {
      "metadata": {
        "colab_type": "code",
        "id": "8nNjvh8G9igc",
        "outputId": "1ac20d3a-1c81-43f5-edf3-200ca7b35542",
        "colab": {}
      },
      "cell_type": "code",
      "source": [
        "!kallisto index -i $WORKDIR/kallisto/transcripts.idx $WORKDIR/microbesOnline/4932.transcriptomes.fasta"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "\n",
            "[build] loading fasta file ./data/microbesOnline/4932.transcriptomes.fasta\n",
            "[build] k-mer length: 31\n",
            "[build] counting k-mers ... done.\n",
            "[build] building target de Bruijn graph ...  done \n",
            "[build] creating equivalence classes ...  done\n",
            "[build] target de Bruijn graph has 11350 contigs and contains 8201152 k-mers \n",
            "\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "metadata": {
        "colab_type": "text",
        "id": "1r30AhLu9ige"
      },
      "cell_type": "markdown",
      "source": [
        "**2b.** Kallisto combines the mapping and quantifying steps into one, so unlike pipeline A which first uses STAR to map the reads to a reference genome sequence and then uses cufflinks to quantify the aligned reads, we can run just one command to get a file representing levels of gene expression."
      ]
    },
    {
      "metadata": {
        "colab_type": "code",
        "id": "IafBa-Fm9igf",
        "outputId": "ee7923e6-f1b6-466d-fa4c-ca5d9adf002a",
        "colab": {}
      },
      "cell_type": "code",
      "source": [
        "!kallisto quant -i $WORKDIR/kallisto/transcripts.idx -o $WORKDIR/kallisto/output \\\n",
        "    $WORKDIR/cleanFASTQ/SRR5511057.forward_paired.fastq $WORKDIR/cleanFASTQ/SRR5511057.reverse_paired.fastq"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "\n",
            "[quant] fragment length distribution will be estimated from the data\n",
            "[index] k-mer length: 31\n",
            "[index] number of targets: 6,333\n",
            "[index] number of k-mers: 8,201,152\n",
            "[index] number of equivalence classes: 7,745\n",
            "[quant] running in paired-end mode\n",
            "[quant] will process pair 1: ./data/cleanFASTQ/SRR5511057.forward_paired.fastq\n",
            "                             ./data/cleanFASTQ/SRR5511057.reverse_paired.fastq\n",
            "[quant] finding pseudoalignments for the reads ... done\n",
            "[quant] processed 6,915,652 reads, 6,505,430 reads pseudoaligned\n",
            "[quant] estimated average fragment length: 159.718\n",
            "[   em] quantifying the abundances ... done\n",
            "[   em] the Expectation-Maximization algorithm ran for 838 rounds\n",
            "\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "metadata": {
        "colab_type": "text",
        "id": "l5aqcJ0r9igi"
      },
      "cell_type": "markdown",
      "source": [
        "Once the above command has finished, you should see a new folder called `output`. If you look inside, there should be a file called `abundance.tsv` with five columns:\n",
        " 1. The first column is `target_id`, which is the identifier of the transcript from the FASTA transcriptome we downloaded. \n",
        " 2. The second column is `length`, which is the length of the transcript in base pairs. \n",
        " 3. The third column is `eff_length`, which stands for effective length. This is the length of the original transcript that matches with reads. \n",
        " 4. The fourth column is `est_counts`, which stands for estimated counts. This is almost the same as _counts_, but accounts for the amount of bias in the experiment. \n",
        " 5. This fifth column is `tpm`, which is the metric for the level of gene expression.\n",
        " \n",
        "For more information about these expression units, visit [this blog post](https://haroldpimentel.wordpress.com/2014/05/08/what-the-fpkm-a-review-rna-seq-expression-units/).\n",
        "\n",
        "\n",
        "**Question:** What would you expect to be the most expressed transcript in yeast to be?<sub>(3)</sub>"
      ]
    },
    {
      "metadata": {
        "colab_type": "text",
        "id": "Xs_zbI7k9igj"
      },
      "cell_type": "markdown",
      "source": [
        "## Extension 1: Evaluating differentially expressed transcripts \n",
        "\n",
        "From this one sample, we can see that some transcripts are expressed more or less than other transcripts, but this data does not give any insight about the effect of changes in transcript abundance and cellular behavior. This is why we need to compare the transcript expression of different samples. For example, imagine we would like to address the following question: which transcripts are responsible for making this a cancerous cell? We would look at the transcript expression between a normal cell and a cancerous cell, and analyze transcripts that are differentially expressed between the two samples. In this extension, we will begin looking at differentially expressed transcripts by visualizing the levels of transcription between different samples.\n",
        "\n",
        "We will continue to use samples from the adaptive prediction paper ([PMID 28854640](https://www.ncbi.nlm.nih.gov/pubmed/28854640)) to analyze the effects of caffeine on yeast at the transcriptome level. \n",
        "\n",
        "Before we download the data from the other samples, let us clean our working directories. For the following folders `$WORKDIR/BAM/`, `$WORKDIR/cleanFASTQ/`, `$WORKDIR/cufflinks/`, `$WORKDIR/FASTQ/`, and `$WORKDIR/kallisto/`, create new folders in each folder specifying the experiment (1a, 1c, 2a, 2c, 3a, 3c). In the pipeline tutorials, we were working with data from 1a, so move any existing files into your new 1a folder. However, in the kallisto folder, leave the transcripts.idx file under /data/kallisto because we can reuse the kallisto index and it should be shared between the samples.\n",
        "\n",
        "For example, your `$WORKDIR/kallisto/` folder should look something like this:\n",
        "<html>\n",
        "<div style=\"width: 50%;margin: 0 auto; background-color: rgb(242, 242, 242);\">\n",
        "    <code><i>kallisto</i>/\n",
        "    transcripts.idx\n",
        "    1a/\n",
        "        abundance.h5\n",
        "        abundance.tsv\n",
        "        run_info.json\n",
        "    1c/\n",
        "    2a/\n",
        "    2c/\n",
        "    3a/\n",
        "    3c/</code>\n",
        "</div>\n",
        "</html>\n",
        "\n",
        "We will now download the sequencing data from replicate 1 after caffeine, replicate 2 before and after caffeine, and replicate 3 before and after caffeine. Research scientists use replicates to check that outcomes are not due to chance. This is different from a trial, because a **trial** refers to a certain action performed many times repeatedly using the exact same object, while a **replicate** would perform a certain action on many of the same object. \n",
        "\n",
        "An example of a trial is flipping the same coin three times, while replicates would be flipping three different coins.\n",
        "\n",
        "\n",
        "| Replicate # | Before Caffeine (a) | After Caffeine (c) |\n",
        "| --- | --- | --- |\n",
        "| 1 | [SRX2785036](https://www.ncbi.nlm.nih.gov/sra/SRX2785036[accn]) | [SRX2785039](https://www.ncbi.nlm.nih.gov/sra/SRX2785039[accn]) |\n",
        "| 2 | [SRX2785037](https://www.ncbi.nlm.nih.gov/sra/SRX2785037[accn]) | [SRX2785040](https://www.ncbi.nlm.nih.gov/sra/SRX2785040[accn]) |\n",
        "| 3 | [SRX2785038](https://www.ncbi.nlm.nih.gov/sra/SRX2785038[accn]) | [SRX2785041](https://www.ncbi.nlm.nih.gov/sra/SRX2785041[accn]) |\n",
        "\n",
        "SRA toolkit provides a very easy to use executable called `prefetch`, which we can call to help us download the sequencing data files. All we have to do is provide the SRA accession, a unique identifier provided by the [NCBI](https://www.ncbi.nlm.nih.gov/) for each sample."
      ]
    },
    {
      "metadata": {
        "colab_type": "code",
        "id": "7uudzND49igk",
        "outputId": "4a64ce68-2f22-466e-80cc-2714915a3176",
        "colab": {}
      },
      "cell_type": "code",
      "source": [
        "!prefetch SRX2785037\n",
        "!prefetch SRX2785038\n",
        "!prefetch SRX2785039\n",
        "!prefetch SRX2785040\n",
        "!prefetch SRX2785041"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "\n",
            "2019-02-21T21:15:09 prefetch.2.9.4: 1) Downloading 'SRX2785037'...\n",
            "2019-02-21T21:15:09 prefetch.2.9.4:  Downloading via https...\n",
            "2019-02-21T21:17:21 prefetch.2.9.4:  https download succeed\n",
            "2019-02-21T21:17:21 prefetch.2.9.4: 1) 'SRX2785037' was downloaded successfully\n",
            "2019-02-21T21:17:21 prefetch.2.9.4: 'SRX2785037' has 0 unresolved dependencies\n",
            "\n",
            "2019-02-21T21:18:37 prefetch.2.9.4: 1) Downloading 'SRX2785038'...\n",
            "2019-02-21T21:18:37 prefetch.2.9.4:  Downloading via https...\n",
            "2019-02-21T21:19:41 prefetch.2.9.4:  https download succeed\n",
            "2019-02-21T21:19:41 prefetch.2.9.4: 1) 'SRX2785038' was downloaded successfully\n",
            "2019-02-21T21:19:41 prefetch.2.9.4: 'SRX2785038' has 0 unresolved dependencies\n",
            "\n",
            "2019-02-21T21:20:57 prefetch.2.9.4: 1) Downloading 'SRX2785039'...\n",
            "2019-02-21T21:20:57 prefetch.2.9.4:  Downloading via https...\n",
            "2019-02-21T21:22:36 prefetch.2.9.4:  https download succeed\n",
            "2019-02-21T21:22:36 prefetch.2.9.4: 1) 'SRX2785039' was downloaded successfully\n",
            "2019-02-21T21:22:36 prefetch.2.9.4: 'SRX2785039' has 0 unresolved dependencies\n",
            "\n",
            "2019-02-21T21:23:52 prefetch.2.9.4: 1) Downloading 'SRX2785040'...\n",
            "2019-02-21T21:23:52 prefetch.2.9.4:  Downloading via https...\n",
            "2019-02-21T21:26:09 prefetch.2.9.4:  https download succeed\n",
            "2019-02-21T21:26:09 prefetch.2.9.4: 1) 'SRX2785040' was downloaded successfully\n",
            "2019-02-21T21:26:09 prefetch.2.9.4: 'SRX2785040' has 0 unresolved dependencies\n",
            "\n",
            "2019-02-21T21:27:25 prefetch.2.9.4: 1) Downloading 'SRX2785041'...\n",
            "2019-02-21T21:27:25 prefetch.2.9.4:  Downloading via https...\n",
            "2019-02-21T21:28:58 prefetch.2.9.4:  https download succeed\n",
            "2019-02-21T21:28:58 prefetch.2.9.4: 1) 'SRX2785041' was downloaded successfully\n",
            "2019-02-21T21:28:58 prefetch.2.9.4: 'SRX2785041' has 0 unresolved dependencies\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "metadata": {
        "colab_type": "text",
        "id": "IBbH6uga9igo"
      },
      "cell_type": "markdown",
      "source": [
        "These files usually download automatically to a new location, the `~/ncbi/public/sra/` folder, so lets move all the files to our /data/seqData folder."
      ]
    },
    {
      "metadata": {
        "colab_type": "code",
        "id": "rbVjzlfJ9igo",
        "outputId": "2f37dcce-50c2-402e-aeea-0934e38621eb",
        "scrolled": true,
        "colab": {}
      },
      "cell_type": "code",
      "source": [
        "!mv -v  ~/ncbi/public/sra/* $WORKDIR/seqData"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "/Users/andrewliu/ncbi/public/sra/SRR5511058.sra -> ./data/seqData/SRR5511058.sra\r\n",
            "/Users/andrewliu/ncbi/public/sra/SRR5511059.sra -> ./data/seqData/SRR5511059.sra\r\n",
            "/Users/andrewliu/ncbi/public/sra/SRR5511060.sra -> ./data/seqData/SRR5511060.sra\r\n",
            "/Users/andrewliu/ncbi/public/sra/SRR5511061.sra -> ./data/seqData/SRR5511061.sra\r\n",
            "/Users/andrewliu/ncbi/public/sra/SRR5511062.sra -> ./data/seqData/SRR5511062.sra\r\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "metadata": {
        "colab_type": "text",
        "id": "2qwOZKx29igq"
      },
      "cell_type": "markdown",
      "source": [
        "Using the same procedure as above, let's generate metrics to measure the level of gene expression in each experiment. Feel free to reference above if you need help, but **remember to change file names and locations**.\n",
        "\n",
        "1. convert to FASTQ\n",
        "2. clean the reads\n",
        "3. map the reads to the reference transcriptome sequence (we already generated the transcriptomeIndex, so no need to do it again!)\n",
        "4. quantify the aligned reads\n",
        "\n",
        "If you get stuck, here is the list of commands to run:"
      ]
    },
    {
      "metadata": {
        "colab_type": "code",
        "id": "YJBqar2E9igr",
        "outputId": "1693347d-88d4-4727-98a7-24cbe08250da",
        "colab": {}
      },
      "cell_type": "code",
      "source": [
        "!fastq-dump --split-files -O $WORKDIR/FASTQ/2a $WORKDIR/seqDATA/SRR5511058.sra"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Read 6811851 spots for ./data/seqDATA/SRR5511058.sra\r\n",
            "Written 6811851 spots for ./data/seqDATA/SRR5511058.sra\r\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "metadata": {
        "colab_type": "code",
        "id": "4dN4VwNw9igt",
        "outputId": "0d56a0c3-ed8e-4788-e6dd-0749f04a9974",
        "colab": {}
      },
      "cell_type": "code",
      "source": [
        "!fastq-dump --split-files -O $WORKDIR/FASTQ/3a $WORKDIR/seqDATA/SRR5511059.sra"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Read 2577860 spots for ./data/seqDATA/SRR5511059.sra\r\n",
            "Written 2577860 spots for ./data/seqDATA/SRR5511059.sra\r\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "metadata": {
        "colab_type": "code",
        "id": "OJL2-KgL9igw",
        "outputId": "f4d680bf-7a86-4886-a677-2737f4c32bbf",
        "colab": {}
      },
      "cell_type": "code",
      "source": [
        "!fastq-dump --split-files -O $WORKDIR/FASTQ/1c $WORKDIR/seqDATA/SRR5511060.sra"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Read 5187380 spots for ./data/seqDATA/SRR5511060.sra\r\n",
            "Written 5187380 spots for ./data/seqDATA/SRR5511060.sra\r\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "metadata": {
        "colab_type": "code",
        "id": "cXQ-PHNa9igy",
        "outputId": "f9e16987-ba86-49a3-e4f9-b48651acf277",
        "colab": {}
      },
      "cell_type": "code",
      "source": [
        "!fastq-dump --split-files -O $WORKDIR/FASTQ/2c $WORKDIR/seqDATA/SRR5511061.sra"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Read 6148895 spots for ./data/seqDATA/SRR5511061.sra\r\n",
            "Written 6148895 spots for ./data/seqDATA/SRR5511061.sra\r\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "metadata": {
        "colab_type": "code",
        "id": "zgT0B_RG9ig0",
        "outputId": "2d8e32c5-f54c-415f-bf7c-23dfc34a69a2",
        "colab": {}
      },
      "cell_type": "code",
      "source": [
        "!fastq-dump --split-files -O $WORKDIR/FASTQ/3c $WORKDIR/seqDATA/SRR5511062.sra"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Read 4211946 spots for ./data/seqDATA/SRR5511062.sra\r\n",
            "Written 4211946 spots for ./data/seqDATA/SRR5511062.sra\r\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "metadata": {
        "colab_type": "code",
        "id": "EI21hNJ99ig2",
        "outputId": "6487edcc-9f4e-4076-bcf2-07417345b61b",
        "colab": {}
      },
      "cell_type": "code",
      "source": [
        "!java -jar ~/software/Trimmomatic-0.38/trimmomatic-0.38.jar PE -phred33 \\\n",
        "    $WORKDIR/FASTQ/2a/SRR5511058_1.fastq $WORKDIR/FASTQ/2a/SRR5511058_2.fastq \\\n",
        "    $WORKDIR/cleanFASTQ/2a/SRR5511058.forward_paired.fastq $WORKDIR/cleanFASTQ/2a/SRR5511058.forward_unpaired.fastq \\\n",
        "    $WORKDIR/cleanFASTQ/2a/SRR5511058.reverse_paired.fastq $WORKDIR/cleanFASTQ/2a/SRR5511058.reverse_unpaired.fastq \\\n",
        "    LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "TrimmomaticPE: Started with arguments:\n",
            " -phred33 ./data/FASTQ/2a/SRR5511058_1.fastq ./data/FASTQ/2a/SRR5511058_2.fastq ./data/cleanFASTQ/2a/SRR5511058.forward_paired.fastq ./data/cleanFASTQ/2a/SRR5511058.forward_unpaired.fastq ./data/cleanFASTQ/2a/SRR5511058.reverse_paired.fastq ./data/cleanFASTQ/2a/SRR5511058.reverse_unpaired.fastq LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36\n",
            "Multiple cores found: Using 4 threads\n",
            "Input Read Pairs: 6811851 Both Surviving: 6381302 (93.68%) Forward Only Surviving: 341875 (5.02%) Reverse Only Surviving: 58767 (0.86%) Dropped: 29907 (0.44%)\n",
            "TrimmomaticPE: Completed successfully\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "metadata": {
        "colab_type": "code",
        "id": "dfRUn51Z9ig4",
        "outputId": "d3f48285-0d5c-4db6-b810-905f363a0d5e",
        "colab": {}
      },
      "cell_type": "code",
      "source": [
        "!java -jar ~/software/Trimmomatic-0.38/trimmomatic-0.38.jar PE -phred33 \\\n",
        "    $WORKDIR/FASTQ/3a/SRR5511059_1.fastq $WORKDIR/FASTQ/3a/SRR5511059_2.fastq \\\n",
        "    $WORKDIR/cleanFASTQ/3a/SRR5511059.forward_paired.fastq $WORKDIR/cleanFASTQ/3a/SRR5511059.forward_unpaired.fastq \\\n",
        "    $WORKDIR/cleanFASTQ/3a/SRR5511059.reverse_paired.fastq $WORKDIR/cleanFASTQ/3a/SRR5511059.reverse_unpaired.fastq \\\n",
        "    LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "TrimmomaticPE: Started with arguments:\n",
            " -phred33 ./data/FASTQ/3a/SRR5511059_1.fastq ./data/FASTQ/3a/SRR5511059_2.fastq ./data/cleanFASTQ/3a/SRR5511059.forward_paired.fastq ./data/cleanFASTQ/3a/SRR5511059.forward_unpaired.fastq ./data/cleanFASTQ/3a/SRR5511059.reverse_paired.fastq ./data/cleanFASTQ/3a/SRR5511059.reverse_unpaired.fastq LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36\n",
            "Multiple cores found: Using 4 threads\n",
            "Input Read Pairs: 2577860 Both Surviving: 2454399 (95.21%) Forward Only Surviving: 99434 (3.86%) Reverse Only Surviving: 16473 (0.64%) Dropped: 7554 (0.29%)\n",
            "TrimmomaticPE: Completed successfully\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "metadata": {
        "colab_type": "code",
        "id": "7GpqfPeA9ig5",
        "outputId": "17795eb4-d0a2-41fb-f9c7-6a16084ef838",
        "colab": {}
      },
      "cell_type": "code",
      "source": [
        "!java -jar ~/software/Trimmomatic-0.38/trimmomatic-0.38.jar PE -phred33 \\\n",
        "    $WORKDIR/FASTQ/1c/SRR5511060_1.fastq $WORKDIR/FASTQ/1c/SRR5511060_2.fastq \\\n",
        "    $WORKDIR/cleanFASTQ/1c/SRR5511060.forward_paired.fastq $WORKDIR/cleanFASTQ/1c/SRR5511060.forward_unpaired.fastq \\\n",
        "    $WORKDIR/cleanFASTQ/1c/SRR5511060.reverse_paired.fastq $WORKDIR/cleanFASTQ/1c/SRR5511060.reverse_unpaired.fastq \\\n",
        "    LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "TrimmomaticPE: Started with arguments:\n",
            " -phred33 ./data/FASTQ/1c/SRR5511060_1.fastq ./data/FASTQ/1c/SRR5511060_2.fastq ./data/cleanFASTQ/1c/SRR5511060.forward_paired.fastq ./data/cleanFASTQ/1c/SRR5511060.forward_unpaired.fastq ./data/cleanFASTQ/1c/SRR5511060.reverse_paired.fastq ./data/cleanFASTQ/1c/SRR5511060.reverse_unpaired.fastq LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36\n",
            "Multiple cores found: Using 4 threads\n",
            "Input Read Pairs: 5187380 Both Surviving: 4894727 (94.36%) Forward Only Surviving: 224109 (4.32%) Reverse Only Surviving: 44706 (0.86%) Dropped: 23838 (0.46%)\n",
            "TrimmomaticPE: Completed successfully\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "metadata": {
        "colab_type": "code",
        "id": "VKCVBAJz9ig7",
        "outputId": "3af69683-e4db-4859-899e-55b52e18878a",
        "colab": {}
      },
      "cell_type": "code",
      "source": [
        "!java -jar ~/software/Trimmomatic-0.38/trimmomatic-0.38.jar PE -phred33 \\\n",
        "    $WORKDIR/FASTQ/2c/SRR5511061_1.fastq $WORKDIR/FASTQ/2c/SRR5511061_2.fastq \\\n",
        "    $WORKDIR/cleanFASTQ/2c/SRR5511061.forward_paired.fastq $WORKDIR/cleanFASTQ/2c/SRR5511061.forward_unpaired.fastq \\\n",
        "    $WORKDIR/cleanFASTQ/2c/SRR5511061.reverse_paired.fastq $WORKDIR/cleanFASTQ/2c/SRR5511061.reverse_unpaired.fastq \\\n",
        "    LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "TrimmomaticPE: Started with arguments:\n",
            " -phred33 ./data/FASTQ/2c/SRR5511061_1.fastq ./data/FASTQ/2c/SRR5511061_2.fastq ./data/cleanFASTQ/2c/SRR5511061.forward_paired.fastq ./data/cleanFASTQ/2c/SRR5511061.forward_unpaired.fastq ./data/cleanFASTQ/2c/SRR5511061.reverse_paired.fastq ./data/cleanFASTQ/2c/SRR5511061.reverse_unpaired.fastq LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36\n",
            "Multiple cores found: Using 4 threads\n",
            "Input Read Pairs: 6148895 Both Surviving: 5871459 (95.49%) Forward Only Surviving: 218792 (3.56%) Reverse Only Surviving: 40469 (0.66%) Dropped: 18175 (0.30%)\n",
            "TrimmomaticPE: Completed successfully\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "metadata": {
        "colab_type": "code",
        "id": "RU803qEH9ihA",
        "outputId": "c55479fd-1f53-4dd6-f0c7-0128247b3080",
        "colab": {}
      },
      "cell_type": "code",
      "source": [
        "!java -jar ~/software/Trimmomatic-0.38/trimmomatic-0.38.jar PE -phred33 \\\n",
        "    $WORKDIR/FASTQ/3c/SRR5511062_1.fastq $WORKDIR/FASTQ/3c/SRR5511062_2.fastq \\\n",
        "    $WORKDIR/cleanFASTQ/3c/SRR5511062.forward_paired.fastq $WORKDIR/cleanFASTQ/3c/SRR5511062.forward_unpaired.fastq \\\n",
        "    $WORKDIR/cleanFASTQ/3c/SRR5511062.reverse_paired.fastq $WORKDIR/cleanFASTQ/3c/SRR5511062.reverse_unpaired.fastq \\\n",
        "    LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "TrimmomaticPE: Started with arguments:\n",
            " -phred33 ./data/FASTQ/3c/SRR5511062_1.fastq ./data/FASTQ/3c/SRR5511062_2.fastq ./data/cleanFASTQ/3c/SRR5511062.forward_paired.fastq ./data/cleanFASTQ/3c/SRR5511062.forward_unpaired.fastq ./data/cleanFASTQ/3c/SRR5511062.reverse_paired.fastq ./data/cleanFASTQ/3c/SRR5511062.reverse_unpaired.fastq LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36\n",
            "Multiple cores found: Using 4 threads\n",
            "Input Read Pairs: 4211946 Both Surviving: 3988920 (94.70%) Forward Only Surviving: 181396 (4.31%) Reverse Only Surviving: 28039 (0.67%) Dropped: 13591 (0.32%)\n",
            "TrimmomaticPE: Completed successfully\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "metadata": {
        "colab_type": "code",
        "id": "_b4Hxklx9ihC",
        "outputId": "81dfa51c-eb0d-48fd-eec1-8866ab9cf1e1",
        "scrolled": true,
        "colab": {}
      },
      "cell_type": "code",
      "source": [
        "!kallisto quant -i $WORKDIR/kallisto/transcripts.idx -o $WORKDIR/kallisto/2a \\\n",
        "    $WORKDIR/cleanFASTQ/2a/SRR5511058.forward_paired.fastq $WORKDIR/cleanFASTQ/2a/SRR5511058.reverse_paired.fastq\n",
        "\n",
        "!kallisto quant -i $WORKDIR/kallisto/transcripts.idx -o $WORKDIR/kallisto/3a \\\n",
        "    $WORKDIR/cleanFASTQ/3a/SRR5511059.forward_paired.fastq $WORKDIR/cleanFASTQ/3a/SRR5511059.reverse_paired.fastq\n",
        "\n",
        "!kallisto quant -i $WORKDIR/kallisto/transcripts.idx -o $WORKDIR/kallisto/1c \\\n",
        "    $WORKDIR/cleanFASTQ/1c/SRR5511060.forward_paired.fastq $WORKDIR/cleanFASTQ/1c/SRR5511060.reverse_paired.fastq\n",
        "\n",
        "!kallisto quant -i $WORKDIR/kallisto/transcripts.idx -o $WORKDIR/kallisto/2c \\\n",
        "    $WORKDIR/cleanFASTQ/2c/SRR5511061.forward_paired.fastq $WORKDIR/cleanFASTQ/2c/SRR5511061.reverse_paired.fastq\n",
        "\n",
        "!kallisto quant -i $WORKDIR/kallisto/transcripts.idx -o $WORKDIR/kallisto/3c \\\n",
        "    $WORKDIR/cleanFASTQ/3c/SRR5511062.forward_paired.fastq $WORKDIR/cleanFASTQ/3c/SRR5511062.reverse_paired.fastq"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "\n",
            "[quant] fragment length distribution will be estimated from the data\n",
            "[index] k-mer length: 31\n",
            "[index] number of targets: 6,333\n",
            "[index] number of k-mers: 8,201,152\n",
            "[index] number of equivalence classes: 7,745\n",
            "[quant] running in paired-end mode\n",
            "[quant] will process pair 1: ./data/cleanFASTQ/2a/SRR5511058.forward_paired.fastq\n",
            "                             ./data/cleanFASTQ/2a/SRR5511058.reverse_paired.fastq\n",
            "[quant] finding pseudoalignments for the reads ... done\n",
            "[quant] processed 6,381,302 reads, 6,107,084 reads pseudoaligned\n",
            "[quant] estimated average fragment length: 158.378\n",
            "[   em] quantifying the abundances ... done\n",
            "[   em] the Expectation-Maximization algorithm ran for 995 rounds\n",
            "\n",
            "\n",
            "[quant] fragment length distribution will be estimated from the data\n",
            "[index] k-mer length: 31\n",
            "[index] number of targets: 6,333\n",
            "[index] number of k-mers: 8,201,152\n",
            "[index] number of equivalence classes: 7,745\n",
            "[quant] running in paired-end mode\n",
            "[quant] will process pair 1: ./data/cleanFASTQ/3a/SRR5511059.forward_paired.fastq\n",
            "                             ./data/cleanFASTQ/3a/SRR5511059.reverse_paired.fastq\n",
            "[quant] finding pseudoalignments for the reads ... done\n",
            "[quant] processed 2,454,399 reads, 2,344,810 reads pseudoaligned\n",
            "[quant] estimated average fragment length: 167.083\n",
            "[   em] quantifying the abundances ... done\n",
            "[   em] the Expectation-Maximization algorithm ran for 913 rounds\n",
            "\n",
            "\n",
            "[quant] fragment length distribution will be estimated from the data\n",
            "[index] k-mer length: 31\n",
            "[index] number of targets: 6,333\n",
            "[index] number of k-mers: 8,201,152\n",
            "[index] number of equivalence classes: 7,745\n",
            "[quant] running in paired-end mode\n",
            "[quant] will process pair 1: ./data/cleanFASTQ/1c/SRR5511060.forward_paired.fastq\n",
            "                             ./data/cleanFASTQ/1c/SRR5511060.reverse_paired.fastq\n",
            "[quant] finding pseudoalignments for the reads ... done\n",
            "[quant] processed 4,894,727 reads, 4,693,261 reads pseudoaligned\n",
            "[quant] estimated average fragment length: 165.64\n",
            "[   em] quantifying the abundances ... done\n",
            "[   em] the Expectation-Maximization algorithm ran for 714 rounds\n",
            "\n",
            "\n",
            "[quant] fragment length distribution will be estimated from the data\n",
            "[index] k-mer length: 31\n",
            "[index] number of targets: 6,333\n",
            "[index] number of k-mers: 8,201,152\n",
            "[index] number of equivalence classes: 7,745\n",
            "[quant] running in paired-end mode\n",
            "[quant] will process pair 1: ./data/cleanFASTQ/2c/SRR5511061.forward_paired.fastq\n",
            "                             ./data/cleanFASTQ/2c/SRR5511061.reverse_paired.fastq\n",
            "[quant] finding pseudoalignments for the reads ... done\n",
            "[quant] processed 5,871,459 reads, 5,663,614 reads pseudoaligned\n",
            "[quant] estimated average fragment length: 152.94\n",
            "[   em] quantifying the abundances ... done\n",
            "[   em] the Expectation-Maximization algorithm ran for 1,057 rounds\n",
            "\n",
            "\n",
            "[quant] fragment length distribution will be estimated from the data\n",
            "[index] k-mer length: 31\n",
            "[index] number of targets: 6,333\n",
            "[index] number of k-mers: 8,201,152\n",
            "[index] number of equivalence classes: 7,745\n",
            "[quant] running in paired-end mode\n",
            "[quant] will process pair 1: ./data/cleanFASTQ/3c/SRR5511062.forward_paired.fastq\n",
            "                             ./data/cleanFASTQ/3c/SRR5511062.reverse_paired.fastq\n",
            "[quant] finding pseudoalignments for the reads ... done\n",
            "[quant] processed 3,988,920 reads, 3,832,481 reads pseudoaligned\n",
            "[quant] estimated average fragment length: 157.058\n",
            "[   em] quantifying the abundances ... done\n",
            "[   em] the Expectation-Maximization algorithm ran for 614 rounds\n",
            "\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "metadata": {
        "colab_type": "text",
        "id": "b-8mfy5p9ihE"
      },
      "cell_type": "markdown",
      "source": [
        "We now have the gene expression in 6 different samples in terms of TPM. A common first step is to validate our data through visualization. We can imagine each sample as a single data point on our graph, where we have n-dimensions for the number of transcripts in yeast. We can expect the \"before-caffeine samples\" to form one cluster and the \"after-caffeine experiments\" to form another cluster. However, how do we transform our high-dimensional transcript expression data into 2-dimensional data?\n",
        "\n",
        "We will explore two popular algorithms for dimensionality reduction: [t-SNE](https://en.wikipedia.org/wiki/T-distributed_stochastic_neighbor_embedding)(t-distributed stochastic neighbor embedding) and [PCA](https://en.wikipedia.org/wiki/Principal_component_analysis)(principal component analysis). \n",
        "\n",
        "We will use Python and use several different libraries. Downloading python libraries is very easy using the [pip](https://pip.pypa.io/en/stable/quickstart/) command."
      ]
    },
    {
      "metadata": {
        "colab_type": "code",
        "id": "duO9MUnX9ihE",
        "scrolled": true,
        "colab": {}
      },
      "cell_type": "code",
      "source": [
        "#import dependencies\n",
        "import pandas as pd\n",
        "from sklearn.manifold import TSNE\n",
        "from sklearn.decomposition import PCA\n",
        "import time\n",
        "import math\n",
        "from plotnine import *\n",
        "import numpy as np\n",
        "import os"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "metadata": {
        "colab_type": "code",
        "id": "QUIOWtWX9ihF",
        "outputId": "3ce61ed0-fba9-46da-85a4-165d0dddd5ce",
        "colab": {}
      },
      "cell_type": "code",
      "source": [
        "#import data\n",
        "a1 = pd.read_csv(os.environ['WORKDIR'] + '/kallisto/1a/abundance.tsv', sep='\\t', header=0, usecols=['target_id', 'tpm'])\n",
        "a2 = pd.read_csv(os.environ['WORKDIR'] + '/kallisto/2a/abundance.tsv', sep='\\t', header=0, usecols=['target_id', 'tpm'])\n",
        "a3 = pd.read_csv(os.environ['WORKDIR'] + '/kallisto/3a/abundance.tsv', sep='\\t', header=0, usecols=['target_id', 'tpm'])\n",
        "c1 = pd.read_csv(os.environ['WORKDIR'] + '/kallisto/1c/abundance.tsv', sep='\\t', header=0, usecols=['target_id', 'tpm'])\n",
        "c2 = pd.read_csv(os.environ['WORKDIR'] + '/kallisto/2c/abundance.tsv', sep='\\t', header=0, usecols=['target_id', 'tpm'])\n",
        "c3 = pd.read_csv(os.environ['WORKDIR'] + '/kallisto/3c/abundance.tsv', sep='\\t', header=0, usecols=['target_id', 'tpm'])\n",
        "\n",
        "#parse each sample's data\n",
        "a1.rename(columns = {'tpm':'a1'}, inplace = True)\n",
        "a2.rename(columns = {'tpm':'a2'}, inplace = True)\n",
        "a3.rename(columns = {'tpm':'a3'}, inplace = True)\n",
        "c1.rename(columns = {'tpm':'c1'}, inplace = True)\n",
        "c2.rename(columns = {'tpm':'c2'}, inplace = True)\n",
        "c3.rename(columns = {'tpm':'c3'}, inplace = True)\n",
        "a2 = a2.drop('target_id', axis=1)\n",
        "a3 = a3.drop('target_id', axis=1)\n",
        "c1 = c1.drop('target_id', axis=1)\n",
        "c2 = c2.drop('target_id', axis=1)\n",
        "c3 = c3.drop('target_id', axis=1)\n",
        "\n",
        "#combine samples into one table, where each row represents a sample and columns represent transcript\n",
        "dataframe = pd.concat([a1,a2,a3,c1,c2,c3], axis=1)\n",
        "dataframe = dataframe.T\n",
        "dataframe.columns = dataframe.iloc[0]\n",
        "dataframe = dataframe.drop('target_id')\n",
        "\n",
        "#clear memory, don't need references anymore\n",
        "del a1,a2,a3,c1,c2,c3\n",
        "\n",
        "#clean data by removing possible noise\n",
        "#filter out columns where any value is less than 10 (lowly expressed)\n",
        "min_threshold = 10\n",
        "for column in dataframe:\n",
        "    for value in dataframe[column]:\n",
        "        if value < min_threshold:\n",
        "            dataframe.drop(column, axis=1, inplace=True)\n",
        "            break\n",
        "\n",
        "#log base 10 all values in dataframe\n",
        "dataframe.applymap(lambda x: math.log(x, 10))"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th>target_id</th>\n",
              "      <th>VIMSS6580644</th>\n",
              "      <th>VIMSS6580645</th>\n",
              "      <th>VIMSS6580646</th>\n",
              "      <th>VIMSS6580647</th>\n",
              "      <th>VIMSS6580648</th>\n",
              "      <th>VIMSS6580649</th>\n",
              "      <th>VIMSS6580651</th>\n",
              "      <th>VIMSS6580653</th>\n",
              "      <th>VIMSS6580656</th>\n",
              "      <th>VIMSS6580657</th>\n",
              "      <th>...</th>\n",
              "      <th>VIMSS6586892</th>\n",
              "      <th>VIMSS6586893</th>\n",
              "      <th>VIMSS6586898</th>\n",
              "      <th>VIMSS6580614</th>\n",
              "      <th>VIMSS6580621</th>\n",
              "      <th>VIMSS6580622</th>\n",
              "      <th>VIMSS6580627</th>\n",
              "      <th>VIMSS6580628</th>\n",
              "      <th>VIMSS6580629</th>\n",
              "      <th>VIMSS6580612</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>a1</th>\n",
              "      <td>2.001253</td>\n",
              "      <td>2.314166</td>\n",
              "      <td>1.591324</td>\n",
              "      <td>1.138631</td>\n",
              "      <td>1.495095</td>\n",
              "      <td>1.540466</td>\n",
              "      <td>1.596420</td>\n",
              "      <td>1.880289</td>\n",
              "      <td>1.723808</td>\n",
              "      <td>2.099069</td>\n",
              "      <td>...</td>\n",
              "      <td>1.387208</td>\n",
              "      <td>1.639635</td>\n",
              "      <td>1.821620</td>\n",
              "      <td>2.330458</td>\n",
              "      <td>1.653391</td>\n",
              "      <td>1.996965</td>\n",
              "      <td>1.362009</td>\n",
              "      <td>1.409023</td>\n",
              "      <td>1.801380</td>\n",
              "      <td>1.902422</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>a2</th>\n",
              "      <td>1.871936</td>\n",
              "      <td>2.351112</td>\n",
              "      <td>1.458062</td>\n",
              "      <td>1.090011</td>\n",
              "      <td>1.459961</td>\n",
              "      <td>1.393377</td>\n",
              "      <td>1.563595</td>\n",
              "      <td>1.805389</td>\n",
              "      <td>1.777433</td>\n",
              "      <td>2.074169</td>\n",
              "      <td>...</td>\n",
              "      <td>1.261981</td>\n",
              "      <td>1.582633</td>\n",
              "      <td>1.857878</td>\n",
              "      <td>2.350058</td>\n",
              "      <td>1.817173</td>\n",
              "      <td>1.938477</td>\n",
              "      <td>1.001145</td>\n",
              "      <td>1.465667</td>\n",
              "      <td>1.930689</td>\n",
              "      <td>1.939737</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>a3</th>\n",
              "      <td>1.955481</td>\n",
              "      <td>2.315626</td>\n",
              "      <td>1.437268</td>\n",
              "      <td>1.211801</td>\n",
              "      <td>1.492533</td>\n",
              "      <td>1.504729</td>\n",
              "      <td>1.595578</td>\n",
              "      <td>1.846698</td>\n",
              "      <td>1.660973</td>\n",
              "      <td>2.140822</td>\n",
              "      <td>...</td>\n",
              "      <td>1.275938</td>\n",
              "      <td>1.631825</td>\n",
              "      <td>1.801223</td>\n",
              "      <td>2.336622</td>\n",
              "      <td>1.524715</td>\n",
              "      <td>2.082688</td>\n",
              "      <td>1.119444</td>\n",
              "      <td>1.549779</td>\n",
              "      <td>1.980582</td>\n",
              "      <td>1.883486</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>c1</th>\n",
              "      <td>1.767758</td>\n",
              "      <td>1.876159</td>\n",
              "      <td>1.507000</td>\n",
              "      <td>1.136264</td>\n",
              "      <td>1.357550</td>\n",
              "      <td>1.263331</td>\n",
              "      <td>1.548698</td>\n",
              "      <td>1.813714</td>\n",
              "      <td>1.963342</td>\n",
              "      <td>2.113850</td>\n",
              "      <td>...</td>\n",
              "      <td>1.284956</td>\n",
              "      <td>1.490526</td>\n",
              "      <td>1.831983</td>\n",
              "      <td>2.325329</td>\n",
              "      <td>1.676088</td>\n",
              "      <td>1.928189</td>\n",
              "      <td>1.233260</td>\n",
              "      <td>1.569479</td>\n",
              "      <td>1.906061</td>\n",
              "      <td>1.853782</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>c2</th>\n",
              "      <td>1.921927</td>\n",
              "      <td>2.002891</td>\n",
              "      <td>1.441230</td>\n",
              "      <td>1.141130</td>\n",
              "      <td>1.332368</td>\n",
              "      <td>1.349048</td>\n",
              "      <td>1.579315</td>\n",
              "      <td>1.799521</td>\n",
              "      <td>1.921784</td>\n",
              "      <td>2.192283</td>\n",
              "      <td>...</td>\n",
              "      <td>1.253885</td>\n",
              "      <td>1.496107</td>\n",
              "      <td>1.850982</td>\n",
              "      <td>2.337631</td>\n",
              "      <td>1.754243</td>\n",
              "      <td>1.941409</td>\n",
              "      <td>1.286061</td>\n",
              "      <td>1.776610</td>\n",
              "      <td>1.915668</td>\n",
              "      <td>1.832984</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>c3</th>\n",
              "      <td>1.895282</td>\n",
              "      <td>1.935375</td>\n",
              "      <td>1.514434</td>\n",
              "      <td>1.191018</td>\n",
              "      <td>1.364104</td>\n",
              "      <td>1.251660</td>\n",
              "      <td>1.655679</td>\n",
              "      <td>1.814967</td>\n",
              "      <td>1.772066</td>\n",
              "      <td>2.164014</td>\n",
              "      <td>...</td>\n",
              "      <td>1.322463</td>\n",
              "      <td>1.495933</td>\n",
              "      <td>1.912814</td>\n",
              "      <td>2.340575</td>\n",
              "      <td>1.693179</td>\n",
              "      <td>1.943146</td>\n",
              "      <td>1.423976</td>\n",
              "      <td>1.710578</td>\n",
              "      <td>1.997486</td>\n",
              "      <td>1.607802</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>6 rows × 4305 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "target_id  VIMSS6580644  VIMSS6580645  VIMSS6580646  VIMSS6580647  \\\n",
              "a1             2.001253      2.314166      1.591324      1.138631   \n",
              "a2             1.871936      2.351112      1.458062      1.090011   \n",
              "a3             1.955481      2.315626      1.437268      1.211801   \n",
              "c1             1.767758      1.876159      1.507000      1.136264   \n",
              "c2             1.921927      2.002891      1.441230      1.141130   \n",
              "c3             1.895282      1.935375      1.514434      1.191018   \n",
              "\n",
              "target_id  VIMSS6580648  VIMSS6580649  VIMSS6580651  VIMSS6580653  \\\n",
              "a1             1.495095      1.540466      1.596420      1.880289   \n",
              "a2             1.459961      1.393377      1.563595      1.805389   \n",
              "a3             1.492533      1.504729      1.595578      1.846698   \n",
              "c1             1.357550      1.263331      1.548698      1.813714   \n",
              "c2             1.332368      1.349048      1.579315      1.799521   \n",
              "c3             1.364104      1.251660      1.655679      1.814967   \n",
              "\n",
              "target_id  VIMSS6580656  VIMSS6580657      ...       VIMSS6586892  \\\n",
              "a1             1.723808      2.099069      ...           1.387208   \n",
              "a2             1.777433      2.074169      ...           1.261981   \n",
              "a3             1.660973      2.140822      ...           1.275938   \n",
              "c1             1.963342      2.113850      ...           1.284956   \n",
              "c2             1.921784      2.192283      ...           1.253885   \n",
              "c3             1.772066      2.164014      ...           1.322463   \n",
              "\n",
              "target_id  VIMSS6586893  VIMSS6586898  VIMSS6580614  VIMSS6580621  \\\n",
              "a1             1.639635      1.821620      2.330458      1.653391   \n",
              "a2             1.582633      1.857878      2.350058      1.817173   \n",
              "a3             1.631825      1.801223      2.336622      1.524715   \n",
              "c1             1.490526      1.831983      2.325329      1.676088   \n",
              "c2             1.496107      1.850982      2.337631      1.754243   \n",
              "c3             1.495933      1.912814      2.340575      1.693179   \n",
              "\n",
              "target_id  VIMSS6580622  VIMSS6580627  VIMSS6580628  VIMSS6580629  \\\n",
              "a1             1.996965      1.362009      1.409023      1.801380   \n",
              "a2             1.938477      1.001145      1.465667      1.930689   \n",
              "a3             2.082688      1.119444      1.549779      1.980582   \n",
              "c1             1.928189      1.233260      1.569479      1.906061   \n",
              "c2             1.941409      1.286061      1.776610      1.915668   \n",
              "c3             1.943146      1.423976      1.710578      1.997486   \n",
              "\n",
              "target_id  VIMSS6580612  \n",
              "a1             1.902422  \n",
              "a2             1.939737  \n",
              "a3             1.883486  \n",
              "c1             1.853782  \n",
              "c2             1.832984  \n",
              "c3             1.607802  \n",
              "\n",
              "[6 rows x 4305 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 26
        }
      ]
    },
    {
      "metadata": {
        "colab_type": "code",
        "id": "KVs1NkW89ihH",
        "outputId": "4146474b-084b-428a-8160-db898c288652",
        "colab": {}
      },
      "cell_type": "code",
      "source": [
        "# perform PCA calculations\n",
        "pca = PCA(n_components=2)\n",
        "pca_result = pca.fit_transform(dataframe.values)\n",
        "\n",
        "print ('Explained variation per principal component {}'.format(pca.explained_variance_ratio_))\n",
        "\n",
        "# perform t-SNE calculations\n",
        "time_start = time.time()\n",
        "tsne = TSNE(n_components=2, verbose=1, perplexity=2, n_iter=5000, learning_rate = 10)\n",
        "tsne_results = tsne.fit_transform(dataframe.values)\n",
        "\n",
        "print ('t-SNE done! Time elapsed: {} seconds'.format(time.time()-time_start))\n",
        "\n",
        "dataframe['PC1'] = pca_result[:,0]\n",
        "dataframe['PC2'] = pca_result[:,1] \n",
        "\n",
        "dataframe['tSNE1'] = tsne_results[:,0]\n",
        "dataframe['tSNE2'] = tsne_results[:,1]\n",
        "\n",
        "# add label column to table to identify each sample\n",
        "dataframe['label'] = pd.Series(['before', 'before', 'before', 'after', 'after', 'after'], index = dataframe.index)"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Explained variation per principal component [0.60461493 0.20639627]\n",
            "[t-SNE] Computing 5 nearest neighbors...\n",
            "[t-SNE] Indexed 6 samples in 0.000s...\n",
            "[t-SNE] Computed neighbors for 6 samples in 0.000s...\n",
            "[t-SNE] Computed conditional probabilities for sample 6 / 6\n",
            "[t-SNE] Mean sigma: 7615.190083\n",
            "[t-SNE] KL divergence after 250 iterations with early exaggeration: 51.728119\n",
            "[t-SNE] KL divergence after 1050 iterations: 0.009043\n",
            "t-SNE done! Time elapsed: 0.17603206634521484 seconds\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "metadata": {
        "colab_type": "text",
        "id": "x_A8Zfma9ihJ"
      },
      "cell_type": "markdown",
      "source": [
        "PCA is a deterministic algorithm that looks for most variant variables across samples. It does this by trying to find a best-fit linear line to the data, which is called the first principal component. The first principal component always accounts for the most variation in the data. Subsequent principal components are the best fitting lines perpendicular to previous principal components. The final points are simply the position of the respective point after being projected onto the principal components. This [video](https://www.youtube.com/watch?v=FgakZw6K1QQ) by StatQuest is very helpful in understanding the step-by-steps of PCA and gaining intuition about how it works. From our example, we can see that 81.1% of the variation in transcript expression in our samples are accounted for by the first and second principal components.\n",
        "\n",
        "t-SNE is a non-deterministic algorithm for dimensionality reduction. Each point is initiated in a position in the wanted lower-dimension. At each step, t-SNE will measure the distances of each point with each other point, and compare these with the similarity values from the original data. The ultimate goal is to make the similarity values between points on the lower dimension to match with the values from the original dimension. This [video](https://www.youtube.com/watch?v=NEaUSP4YerM) by StatQuest helps explain t-SNE. Keep in mind that t-SNE is non-deterministic, meaning that each run will behave differently and result in different results.\n",
        "\n",
        "Note: to re-run t-SNE calculations, you have to first reload the data in the previous code cell before running code cell with t-SNE calculations"
      ]
    },
    {
      "metadata": {
        "colab_type": "code",
        "id": "YYzhmx929ihJ",
        "outputId": "6a8206eb-11de-46a0-fc98-6fc67d512c2f",
        "colab": {}
      },
      "cell_type": "code",
      "source": [
        "#graph PCA \n",
        "chart = ggplot(dataframe, aes(x='PC1', y='PC2', color='label') ) \\\n",
        "        + geom_point(size=3,alpha=1) \\\n",
        "        + ggtitle(\"First and second principal component\")\n",
        "chart"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAArYAAAHICAYAAAC7w6nAAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAAPYQAAD2EBqD+naQAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAIABJREFUeJzs3Xl8E3X+x/F3kjbpldJCy1EKLceCyC0CCoICFYEVBRFQAReUy1t0VUQ8gdVVxANQEF1BQHARUfBEEFDQZV1FRGAVsFCsXIVKS+8k8/uDX7PEttCW0DTj6/l48HjQmW9mPjOftH138s3EYhiGIQAAACDIWQNdAAAAAOAPBFsAAACYAsEWAAAApkCwBQAAgCkQbAEAAGAKBFsAAACYAsEWAAAApkCwBQAAgCkQbAEAAGAKBFsERGJiolJSUgJdRsDs3r1bFotFU6dODXQp1Y7L5ZLFYtHo0aOrfN9V9bys6v7zfAPwR0Gwhd+sX79eFoulzH+LFi065zW88847euKJJ875fgDgXJkxY4beeOONQJcBBKWQQBcA87n22mt19dVXl1jetWtX7//37Nkjq9X/f1e98847Wrp0qR555BG/bxvmd66el7/XpEkT5eXlKTQ09JzvC8FnxowZOu+883TjjTcGuhQg6BBs4Xdt27bV8OHDTzvG4XCUa1snTpxQVFSUP8oCSlVUVCSPxyOHw1Hu5+XZslgsCgsLq5J9AcAfCVMREBClzWUsXvbdd9+pX79+iomJUWxsrCTJ4/HohRdeULt27RQdHa3IyEglJyfrhhtu0JEjR7yPX7x4sdxud4WmQOzcuVO33nqrWrZsqejoaIWHh6tNmzZ67rnn5PF4fMa++uqrslgs+vzzz/X000+radOmcjgcatKkiV588cVSt79o0SK1adNGYWFhql+/vv76178qLy+v3OcqPz9fjz/+uFq0aKHIyEg5nU796U9/0s0336yioiKfsd98840GDRqk+Ph42e12NWnSRJMmTSp1f4cPH9aECRO8xxAXF6fu3btr2bJlPuP279+vm2++WfXr15fdbleDBg00fvx4HTp0yGfcmjVrvOf7tddeU6tWreRwOJSYmKiHH35Ybre7RA2ffPKJOnfurPDwcNWuXVujR4/WsWPHyn1uive5cOFCvfjii2revLnCwsLUuHFjTZs2rcQ+hw8frpCQEB05ckRjxoxRvXr15HA49PXXX0s6/fNyx44d6tevn6Kjo+V0OtW/f3+lpqaWqKmoqEjPPvusLrjgAkVGRio6Olpt27bV448/7h1T2pzXU5ctWbJE7dq183nO5Obm+uznl19+0b333qv27durZs2aCgsLU/PmzfXwww8rPz+/3OewNIZh6B//+Ie6dOni/X4777zzdPfdd8vlcnnHeTwePffcc2rTpo3Cw8NVo0YNpaSkaO3atSW2WXwet27dqt69e8vpdKpWrVoaO3ascnNz5fF49NRTT6lJkyZyOBxq2bKlPvjgA59tVPQcSef2+btr1y7deOONSkhI8G77tttuK/Ecnjx5siwWi3bv3q2JEyeqQYMG3mN86623Shxfenq61q5d6/Nz7Jdffilf84A/OK7Ywu9yc3OVkZHhsyw0NFQ1atQ442P37t2ryy67TNdcc43+/ve/e3/5TJkyRY899pj69++vMWPGKDQ0VGlpafroo490+PBhxcfHa+bMmZo+fbr+9a9/acGCBd5tnjoFojRr167Vxo0bdfXVVyspKUm5ubl6//33dc899yg1NbXUwHr//fcrJydHo0ePVmRkpObPn6+77rpLCQkJuvbaa73jZs2apTvuuEPNmzfXY489JpvNpkWLFmn9+vVnPBfFbrnlFs2fP18jRozQXXfdJcMwlJqaqlWrVqmgoMD7cvYHH3yga665RsnJybrrrrtUu3ZtbdmyRdOnT9eXX36pNWvWKCQkxHueu3btqoMHD2r48OHq2LGjCgoK9O233+r999/X4MGDJUlpaWnq1KmTjh49qrFjx6pVq1basmWL5s2bp48//lhff/214uPjfeqdNWuWDhw4oNGjR6tWrVpavny5pk6dqho1auivf/2rd9x7772na665RnXr1tXEiRMVExOj5cuXq1+/fuU+N8Wef/55/frrrxo3bpxiY2O1YsUKTZ48WampqXr11Vd9xhqGoZSUFNWuXVsTJ06Ux+NRnTp1Trv9/fv367LLLtPAgQP1zDPP6Mcff9SsWbN09dVXa+vWrbJYLJKkwsJCXXHFFVq/fr169uypRx99VFFRUdq5c6fefvttPfroo2c8lhUrVmjXrl267bbbNGbMGK1evVrPPvusvv/+e33yySfefX333Xd69913NXDgQDVq1Egej0efffaZpk6dqq1bt2rlypUVPo/F52fEiBFavHixLrzwQj3wwAOKi4vT7t279c4772jq1KneV1FGjBihN998U927d9eTTz6prKwsvfrqq+rdu7cWL16s6667zmfbaWlpSklJ0dChQzVo0CB9+eWXmjdvngoKChQeHq5///vfGj9+vGw2m5577jldc8012r17txo0aFCpc3Qun7/ffPONevbsKafTqZtvvlkNGzbUjz/+qDlz5mjdunXavHmznE6nz7aL/7C6++67ZbFY9NJLL+m6665TkyZNdOGFF6pu3bpauHCh7rzzTiUkJGjixInex9asWbNS/QT+cAzAT9atW2dIKvVf586dfcbWr1/f6NWrV4llkox58+aV2Hbr1q2N1q1bn7GGYcOGGTabrUJ1nzhxotTlQ4cONUJCQoxDhw55l82bN8+QZLRv394oKCjwLs/KyjJiY2ONbt26eZcdO3bMiIiIMBo1amQcP37cuzwnJ8do3bq1IcmYMmXKGeuLjo42+vfvf9oxOTk5Rnx8vHHxxRf71GUYhrF06VJDkrFo0SLvsiuuuMKQZLz//vsltuV2u73/v/766w1JxvLly33GvPbaa4YkY9y4cd5ln376qSHJSEhIMDIzM32216xZM6NBgwbeZS6Xy2jQoIERHR1tpKen+ywvru3mm28+7TGfus+oqCgjLS3NZzt9+/Y1JBlfffWVd/mwYcMMScbw4cNL3d7pnpf//Oc/fZZPmTLFkGSsXbvWu+zJJ580JBn33XdfiW2fel537dpVov/Fy6xWq/HNN9/4PHbcuHGGJGPJkiXeZbm5uYbH4ymxnwceeMCQZHz77ben3V9ZlixZYkgyrrvuOsPlcvmsO3V/n3zyiSHJuPrqq32O7eDBg0ZcXJxRs2ZNn++t4vO4bNkyn23279/fsFgsRocOHYzCwkLv8q+//tqQZEyePLnS5+hcPX89Ho/RunVro0mTJj5jDcMwNm3aZFitVmPq1KneZQ899JAhybjyyit9zlVqaqoREhJijBgxwmcbpT0PAZQPUxHgdyNHjtSnn37q82/mzJnlemx8fLxuuummEstjY2OVlpamjRs3+rtcRUZGev9fUFCgY8eOKSMjQ3369JHL5dK3335b4jF33HGH7Ha792un06nOnTvrp59+8i77+OOPlZubqzvuuEPR0dHe5REREbr33nvLXV9MTIy+//57bdu2rcwxq1ev1pEjRzRq1ChlZWUpIyPD+69Hjx4KCwvTJ598Ikk6cuSIVq9erZSUFP35z38usa3iN0+5XC6tXLlSLVu21DXXXOMzZuTIkUpKStLy5ctLPH706NGKiYnx2V7Pnj21f/9+75SIf//739q/f79GjhyphIQE71ibzaYHH3yw3Oem2I033uhzVc9ms3mvdpVW4wMPPFCh7Tds2NB7FbvY5ZdfLkk+PV+0aJGio6N9ph0UK++b0vr06aMLLrjAZ9mkSZMk+R5LeHi498pkUVGR93lbXNfmzZvLtb/fK566M336dNlsNp91xfs7tZaHH37Y59jq1KmjW265RceOHSvxykTDhg19XtGQpG7duskwDN12220+b6a78MILFRERoV27dpWosTzn6Fw+f7du3apt27bp+uuvl8vl8vl+a9asmRo1auT9fjvVPffc43OukpOT1bRpU5/nEICzQ7CF3zVp0kQpKSk+/zp27FiuxzZt2rTUAPD0008rPDxc3bp1U7169TR48GDNnTtXWVlZZ11vTk6OHnjgASUnJyssLEy1atVSfHy8Ro0aJUmlzvls3LhxiWW1atXS0aNHvV/v2bNHknT++eeXGNuyZcty1zdr1ixlZmaqTZs2Sk5O1ogRI7Ro0SIVFBR4x+zcuVOSNHbsWMXHx/v8q1OnjvLz873TOnbt2iXDMNS+ffvT7vfQoUPKyclRq1atSqyzWq1q2bKlMjIySvSgrHMj/e9c+uvcFCttO8XLdu/eXWLdn/70pwpt/3THdGrPf/rpJzVv3lzh4eEV2v6pSjuWhg0bKioqyudYXC6XnnzySZ133nlyOBze523xHOGKzFU+1U8//aT4+HjVr1//tON+/vnnMutt3bq1pP/1uVhp57F4Hn1Z6049v8XKc47O5fO3+Ptt6tSpJb7f4uPjtWfPnhJzeE+37dKOEUDlMMcW1UpERESpyzt37qw9e/Zo9erVWr9+vdavX6+3335bjzzyiDZt2qSmTZtWep9Dhw7Vhx9+qDFjxqh79+6qVauWQkJC9PXXX2vSpEkl3kAmqcSVrGKGYZRrn+UdJ0n9+/fXvn379NFHH2nDhg1at26dFi1apCeeeEJfffWVatWq5a3xqaeeUocOHUrdTvEv54rWeOpVuvIo69yUd98VOTfl8fv6bTZbhe9+cLbH5A+GYfgcy913363Zs2dr6NChmjhxomrXri273a60tDTdfPPNpT5vy7ufytRTHqc7j2f7PfX7ms7l87f43N59992lvuohlf6zzB/HCOD0CLYIGhERERowYIAGDBggSd43zjzzzDOaO3eupIr/Ejt69Kg++OADjRo1yruNYsVXZSqrSZMmkqQdO3boiiuu8Fm3Y8eOCm0rJiZG119/va6//npJJ98sNWHCBL3yyit68MEH1axZM0knX54+0ydnNWvWTBaLRVu2bDntuLp16yoyMlI//PBDiXWGYWjHjh2Kj4/3mWZRXqeem9+r6Lk503aK91UVmjVrph9//FF5eXmVvmpb2rGkpaUpJyfH51jeeOMN9ejRQ0uXLvUZu2rVqkrtt1jz5s31wQcfKD09/bRXbZs2baq1a9dqx44dJf6YKn7OnKtzX55zdC6fv8XfbxaL5Zx8Ul1Ff44B+B+mIiAoFN/S61TFv0xPfck1KipKbre73FMUiq+g/P6KSVZWlp5//vnKlivp5DzA8PBwzZw506eevLw8Pfvss+Xahsvl0m+//VZi+e+PvW/fvoqPj9ff//53HT58uMT4oqIiZWZmSjo5j7l3795as2ZNidspSf+7GhUSEqKrrrpKP/zwg959912fMQsWLNDevXtLzF0sr44dOyoxMVHz58/Xr7/+6l3udrv15JNPVnh7b7zxhvbv3+9zDE899ZQkVbrGyhg+fLiysrJKvftBea+gfvzxxyXmdf/tb3+T5HssNputxPO2sLCwUufvVMX3oL733ntL3OLq1P0V1zJt2jSfYzty5Ihefvll1axZUz169DirWspSnnN0Lp+/HTp0UMuWLfXKK6/oxx9/LLHe4/GU+jOrvKKioio9lQT4o+OKLYJCs2bN1LVrV3Xq1En169fX0aNH9frrr8tisegvf/mLd9xFF12kOXPmaPz48erbt69CQ0N18cUXKykpqdTtxsTE6PLLL9eCBQsUGhqqTp06KT09Xa+99prq1q3rnUdYGbGxsXryySd19913q3Pnzho5cqRsNpsWLlzo88az0/ntt9/UoEED9e/fX+3atVPdunWVnp6uuXPnym6364YbbpB08hfhG2+8oYEDB+q8887TTTfdpGbNmik7O1u7du3SO++8oxkzZnhDy8svv6wuXbroqquu0vDhw9WpUyfvG+UsFovmz58v6eTUhs8++0xDhgzx3i7pu+++07x585SUlKQpU6ZU6tyEhIToxRdf1LXXXquOHTtq3LhxqlGjhpYvX17qvUjPpFmzZurcubP3dl/vvPOONmzYoJEjR+riiy+uVI2Vcc899+jDDz/UM888o2+++UZ9+vSR0+nUjz/+qLVr1+r7778/4zbatWunHj166NZbb1ViYqJWr16tlStXqmfPnho6dKh33ODBgzVv3jwNGTJEKSkpOnr0qBYtWuTzZsjKGDp0qN577z0tXbpUe/bs0cCBAxUXF6fU1FQtW7ZM3333naKiotS7d29df/31WrJkiXr16qUBAwZ4b/d19OhRLV68uMypRWervOfoXD1/rVarFi5cqF69eql9+/YaNWqUWrVqpcLCQv3888969913NWbMGE2ePLlS27/ooou0YMECPfroo2revLmsVquuvvrqs5q7DfxhVOUtGGBuxbf7Ks8thcq6rVJZt7iZNm2a0b17d6N27dpGaGioUa9ePePPf/6z8dlnn/mMc7lcxoQJE4yEhATDarUakoyFCxeetpYjR44Yo0ePNhISEgyHw2Gcd955xvTp0723Mzr18cW3+/riiy9KbKesW40tWLDAaNWqlWG3242EhATjnnvuMbZu3Vquc5Wfn29MnDjR6NSpk1GrVi3DbrcbDRo0MIYMGVLidkeGYRjbt283brzxRqN+/fpGaGioERcXZ3To0MGYNGmSsX//fp+xv/76q3HrrbcaSUlJ3rGXXnqp8fbbb/uMS0tLM0aNGmXUq1fPCAkJMRISEoyxY8caBw4c8BlXfLuk0s538e2Ofl/DBx98YHTs2NFwOBxGXFyccdNNNxkHDx6s8O2+3njjDeOFF14wmjVrZtjtdiMpKcl44oknStyu6ky3g6vI87KsW2jl5+cbTz75pNGqVSvD4XAYTqfTaNu2rc/tn053u68pU6YYb775ptGmTRvD4XAY9erVMyZMmFDitnQ5OTnGfffdZyQlJRl2u91o1KiR8dBDDxnbtm077bbLw+12Gy+//LJx4YUXGuHh4UZkZKTRokULY8KECUZRUZF3nMvlMmbMmOFzrD179jQ+/fTTcp1bwzj999TvH1PRc2QY5/b5m5aWZtx6661Go0aNDLvdbsTGxhpt2rQxJkyYYOzcufOMjzcMw+jatavRpEkTn2UHDhwwBg4caMTExBgWi6XMxwIoyWIYzFoHEJzWrFmjyy+/XAsXLjzjxzhXd7t379af/vQnTZkypdJX+syOcwTgTJhjCwAAAFMg2AIAAMAUCLYAAAAwBebYAgAAwBS4YgsAAABTINgCAADAFAi2AAAAMAWCLQAAAEyBYAsAAABTINgCAADAFEICXUAwy83NVV5ensx2xzS73a7CwsJAl+E3FotF4eHhpusVfQoO9Ck40Kfg4K8+xcXF+aEaVEdcsT0LERERslrNdwodDkegS/Arq9Vqyl7Rp+BAn4IDfQoOZusT/M9cz3gAAAD8YRFsAQAAYAoEWwAAAJhCUL157NChQ5o7d67++9//ymazqWPHjho7dqzCwsJ05MgRzZw5Uzt37lRsbKxGjhypLl26eB+7adMmzZ8/X7/99ptatGihu+66S7Vq1ZIkFRUV6ZVXXtHGjRtls9nUt29fDRs2LFCHCQAAgEoIqiu2L730kqKiovT6669r9uzZSk9P11tvvSVJmj59uurXr69Fixbplltu0QsvvKD09HRJ0i+//KIXX3xRt912mxYtWqSEhARNnz7du9233npLe/fu1dy5c/Xss89qw4YNWrNmTUCOEQAAAJUTVMH24MGD6t69uxwOh6Kjo3XRRRdp3759+vXXX7Vr1y4NGzZMDodD7du3V7t27bRu3TpJ0rp163TBBReoXbt2cjgcGjZsmP773//qwIEDkqQ1a9Zo6NChio6OVp06dXTVVVcRbAEAAIJMUE1FuOqqq7Rhwwa1atVK+fn5+uqrr3TZZZdp3759io+PV1RUlHdso0aN9PPPP0uS9u3bp2bNmnnXOZ1OxcfHa9++fXI6nTp27JgaNWrk89jFixeX2H9GRoYyMjK8X8fFxclut5+LQw0oi8Uim80W6DL8pvhYzHRMEn0KFvQpONCn4GC2PsH/girYtm7dWmvXrtV1110nj8ejjh07qnfv3vriiy8UGRnpMzYqKkp5eXmSpPz8fEVERPisj4yMVF5ennfMqeuL1/3e8uXLNW/ePO/XY8aM0bhx4/x2fNWJGQN7dHR0oEvwO/oUHOhTcKBPwcGMfYL/BE2wdbvdeuyxx5SSkqKnnnpKLpdL8+bN04wZM9StWzfl5ub6jM/JyVF4eLgkKSwsrMz1xWNyc3NL/f+pBg0apEsvvdT7dVxcnLKysuR2u/16rIEWGRmpnJycQJfhNzabTdHR0abrFX0KDvQpONCn4OCvPsXGxvqhGlRHQRNsc3JylJGRoSuvvFJ2u112u139+vXTQw89pOHDh+vw4cPKycnxXrlNTU1Vw4YNJUlJSUnau3evd1snTpxQRkaGkpKSFBUVpZo1a2rv3r3euyT8/PPPSkpKKlFDXFxciY/hy8zMNNUPDUkyDMN0xySd/OPITMdFn4IDfQoO9Ck4mLVP8J+gefNYdHS06tatqw8//FBFRUXKz8/XJ598ouTkZCUkJKhp06Z68803VVBQoK1bt2rLli3q0aOHJOmyyy7TN998o61bt6qgoECLFy9W8+bNVa9ePUlSr1699NZbbykrK0uHDh3SypUrlZKSEsjDBQBT8uTlquCL9Trxjzk68doc5X/+mTy/e0UNACrLYhiGEegiyis1NVWvvfaa901h5513nsaMGaN69erpyJEjevHFF733sR01apTPfWw3btyoBQsWKDMzU+eff36p97H94osvvPexHT58eLlqMuMVW6fTqezs7ECX4Tc2m02xsbGm6xV9Cg706X+Kdv+knH/Mkdzuk/9OblCyWhU5coxCm59/DiouH/oUHPzVp9+/+grzCKpgWx2Z7YeGxA/4YEGfggN9Osl9NEPZz0yVXK6yNiznXyfJFl/HT5VWDH0KDgRbnEnQTEUAAASvgk2fS2e4jlLwxfqqKQaAaRFsAQDnXNH3W/43/aA0breKtm6puoIAmBLBFgBw7hUVnXGIUdY0BQAoJ4ItAOCcs9ZLkCyW046x1a1XRdUAMCuCLQDgnAu75DJJpwm2Vqsc3XpUVTkATIpgCwA450JatlboBR0kq63kSqtNoa3aKLRNu6ovDICpBM0njwEAgpfFYlHE0BEqbJCs/PVrZPyWeXJ5jRg5Lu0pxyWXyWLlWguAs0OwBQBUCYvVKscll8repZuM7KyTy5zRBFoAfkOwBQBUKYvVKkuNmECXAcCE+DMZAAAApkCwBQAAgCkQbAEAAGAKBFsAAACYAsEWAAAApkCwBQAAgCkQbAEAAGAKBFsAAACYAsEWAAAApkCwBQAAgCkQbAEAAGAKBFsAAACYAsEWAAAApkCwBQAAgCkQbAEAAGAKBFsAAACYAsEWAAAApkCwBQAAgCkQbAEAAGAKBFsAAACYAsEWAAAApkCwBQAAgCkQbAEAAGAKBFsAAACYAsEWAAAApkCwBQAAgCkQbAEAAGAKBFsAAACYAsEWAAAApmAxDMMIdBHBrLCwUGY7hSEhIXK5XIEuw28sFovsdrvpekWfggN9Cg70KTj4q08Oh8MP1aA6Cgl0AcEuJydHbrc70GX4ldPpVHZ2dqDL8BubzSa73W66XtGn4ECfggN9Cg7+6hPB1ryYigAAAABTINgCAADAFAi2AAAAMAWCLQAAAEyBYAsAAABTINgCAADAFAi2AAAAMAWCLQAAAEyBYAsAAABTINgCAADAFAi2AAAAMAWCLQAAAEyBYAsAAABTINgCAADAFAi2AAAAMAWCLQAAAEyBYAsAAABTINgCAADAFAi2AAAAMAWCLQAAAEyBYAsAAABTINgCAADAFAi2AAAAMAWCLQAAAEyBYAsAAABTINgCAADAFAi2AAAAMAWCLQAAAEyBYAsAAABTINgCAADAFAi2AAAAMAWCLQAAAEyBYAsAAABTINgCAADAFAi2AAAAMAWCLQAAAEyBYAsAAABTINgCAADAFEICXUBlZGVl6ZZbblG9evU0ffp0SdK+ffs0c+ZM7d27V3Xr1tUtt9yili1beh/z/vvv6+2331ZeXp46dOig22+/XREREZKkEydOaPbs2fr2228VHh6uIUOGqF+/fgE5NgAAAFROUF6x/cc//qHk5GTv1y6XS1OnTtXFF1+sJUuWaNCgQZo2bZpOnDghSdqyZYuWLl2qRx55RK+//rqKioo0d+5c7+Pnzp0rt9ut119/XQ8//LAWL16s77//vqoPCwAAAGch6ILttm3bdPDgQfXo0cNnWUFBgQYOHKjQ0FD16NFDderU0ZdffilJ+uyzz9SrVy81btxYERERGjZsmDZu3KiCggLl5+dr06ZNGj58uCIiItSkSRP17NlTa9asCdQhAgAAoBKCKtgWX2kdP368LBaLd3laWpqSkpJktf7vcBo1aqS0tDRJJ6cpNGrUyLsuKSlJHo9Hv/76q9LT0yVJDRs29K5v3Lix9u3bd64PBwAAAH4UVHNsly1bpvbt2ys5OVl79uzxLs/Ly1NkZKTP2MjISOXm5kqS8vPzfdZbLBZFREQoLy9PhmEoPDy8xGPz8vJK7D8jI0MZGRner+Pi4mS32/1ybNWJxWKRzWYLdBl+U3wsZjomiT4FC/oUHOhTcDBbn+B/QRNsf/31V61fv14vvPBCiXXh4eHeEFssNzfXG1jDwsLKXO/xeEqE2JycnBJhV5KWL1+uefPmeb8eM2aMxo0bV+ljqs7MGNijo6MDXYLf0afgQJ+CA30KDmbsE/wnaILtzp07dfToUY0ePVrSyTeMFRYWatiwYbrjjju0b98+eTwe73SE1NRU9enTR9LJqQepqam69NJLJZ2cmmC1WpWQkCDDMCRJ+/fvV4MGDbyPTUpKKlHDoEGDvNuQTl6xzcrKktvtPncHHgCRkZHKyckJdBl+Y7PZFB0dbbpe0afgQJ+CA30KDv7qU2xsrB+qQXUUNMH2kksuUdu2bb1fb9q0SevWrdPkyZMVExOj0NBQvfvuu+rfv7++/PJLHTx4UBfp2tV6AAAgAElEQVRffLEkqWfPnpoxY4YuvfRS1alTR4sXL9Yll1wih8MhSeratasWL16sO++8U4cOHdLatWt1//33l6ghLi5OcXFxPssyMzNN9UNDkgzDMN0xSZLb7TbVcdGn4ECfggN9Cg5m7RP8J2iCrcPh8AZRSYqKilJISIg3aE6ePFmzZs3Sm2++qTp16mjSpElyOp2SpPbt22vo0KF6/PHHlZubqw4dOvhMIRg3bpxmzZqlkSNHeu+acGqIBgAAQPVnMYpfi0elmPGKrdPpVHZ2dqDL8BubzabY2FjT9Yo+BYc/Qp+MokIV/XeHjOxsWWNiFNL8fFmC7A0+f4Q+mYG/+vT7V19hHkFzxRYAUL0YhqHCTRuU99EqqahIstokj1uWsHCFD7hW9gs6BrpEAH8wBFsAQKUUfLFe+avekYpf+PN4JElGbo5yl7whWSyyt78wgBUC+KMJqg9oAABUD0ZBvvI/Wvm/UFtigKG8lctl/H/YBYCqQLAFAFRY4bat3iu0ZTFOnJBrz64qqggACLYAgErwZB2XLGf4FWKzycg6XjUFAYAItgCASrA6a0jGGaYZuN2yOM33yVcAqi+CLQCgwuyt257xiq0lMkohTZtVUUUAQLAFAFSCJSxMYX2vlCyWMgZYFH7VNbJY+TUDoOpwuy8AQKU4uveUxWpV3kfvSy6XZLWevI+tI4z72AIICIItAKBSLBaLHN16yN65q4p2bpdx4pRPHgvh1wuAqsdPHgDAWbHY7bK3bR/oMgCAObYAAAAwB4ItAAAATIFgCwAAAFMg2AIAAMAUCLYAAAAwBYItAAAATIFgCwAAAFMg2AIAAMAUCLYAAAAwBYItAAAATIFgCwAAAFMg2AIAAMAUCLYAAAAwBYItAAAATIFgCwAAAFMg2AIAAMAUCLYAAAAwBYItAAAATIFgCwAAAFMg2AIAAMAUCLYAAAAwBYItAAAATIFgCwAAAFMg2AIAAMAUCLYAAAAwBYItAAAATIFgCwAAAFMg2AIAAMAULIZhGIEuIpgVFhbKbKcwJCRELpcr0GX4jcVikd1uN12v6FNwoE/BgT4FB3/1yeFw+KEaVEchgS4g2OXk5Mjtdge6DL9yOp3Kzs4OdBl+Y7PZZLfbTdcr+hQc6FNwoE/BwV99ItiaF1MRAAAAYAoEWwAAAJgCwRYAAACmQLAFAACAKRBsAQAAYAoEWwAAAJgCwRYAAACmQLAFAACAKRBsAQAAYAoEWwAAAJgCwRYAAOAceuyxxxQVFVWhxyQnJ+v2228P2P6DFcEWAAAApkCwBQAAgCkQbAEAAKpITk6Obr/9djVv3lwRERFKTk7W+PHjdfz48VLHP/PMM6pfv74iIiJ09dVX68CBAz7rCwoKNGnSJCUlJcnhcKhFixZ68803q+JQqqWQQBcAAADwR5Gbmyu3261p06YpPj5e+/fv17Rp0zRw4EB99tlnPmNXrFihpKQkvfzyy8rMzNTEiRN1zTXX6KuvvvKOGTJkiDZu3KhHH31ULVq00Icffqjhw4crNjZWffv2rerDCziCLQAAQBWJj4/Xyy+/7P3a5XKpUaNGuuSSS/TTTz+pWbNm3nXZ2dn68MMPFRMTI0lKTExUSkqKVq9erd69e2vdunVauXKlPvnkE/Xu3VuSdPnllys9PV2PPvroHzLYMhUBAACgCi1cuFDt27dXVFSUQkNDdckll0iSfvrpJ59xPXr08IZaSerVq5eio6P1r3/9S5K0evVq1axZUz179pTL5fL+69Wrl7Zs2SK32111B1VNcMUWAACgiqxYsUI33nijxo4dq2nTpqlWrVo6cOCABg4cqPz8fJ+xtWvXLvH42rVre+fZZmRk6NixYwoNDS11XwcOHFBiYqL/D6IaI9gCAABUkWXLlqldu3aaO3eud9mGDRtKHXv48OFSl9WrV0+SVLNmTcXHx+vDDz8s9fGlBWOzI9gCAABUkby8PNntdp9lixcvLnXsunXrdPz4cdWoUUOStHbtWmVlZalz586SpJSUFD399NOy2+1q06bNuS08SBBsAQAAqsjll1+u2267TU888YS6dOmijz76SGvXri11rNPpVN++fTVx4kT99ttveuCBB9SpUyddccUV3m31799fffr00f333682bdooJydH27dv1+7du/Xqq69W5aFVCwRbAACAKjJu3Dj9/PPPmjVrlqZPn64rrrhCb775pi666KISYwcOHKjExESNHz9emZmZSklJ8ZnCIElvv/22nnrqKb300kvat2+fatSooVatWmnUqFFVdUjVisUwDCPQRQSzzMxM073r0Ol0Kjs7O9Bl+I3NZlNsbKzpekWfggN9Cg70KTj4q09xcXF+qAbVEbf7AgAAgCkQbAEAAGAKBFsAAACYAm8eAwAAOEfO1dxtp9N5TrYb7LhiCwAAAFMg2AIAAMAUKhVsi4qKdOjQIZV1p7Ds7Gx9/vnnZ1UYAAAAUBEVCraGYeiBBx5QTEyMEhISVLt2bT355JMl7pG3Y8cO9ejRw6+FAgAAAKdToWA7d+5cPffccxo/frwWLFiggQMH6tFHH1WPHj2UmZl5rmpENXHQ5dbanHx9fCJfP+QXycNnewAAEDReeeUV1atXT1FRUfr5558DXc45UaG7Irz88suaNGmSHnvsMUnS8OHDNXbsWA0aNEjdunXTxx9/rMTExHNRJwIo1+PR/N9ytaPQ5X3CuCVFWy26KSZSTezcXAMAgMoyiopUsGObPFlZskZHy3F+a1lCQ/26j6KiIt111136/PPP1bFjR0mSxWLRzp07dd555/l1X4FUoUSyZ8+eElMMLrzwQm3evFl9+/bVRRddpE8++cSvBSKwPIah2cdy9Ivr5HQT1ynrsjyGZh47oftrOZUQagtMgQAABLHcL7/QiRXLZBQWSDab5HbLYncoauBgRXTp5rf9HDx4UPn5+WrdurXftllUVKRQPwfws1WhqQg1a9bUoUOHSiyvW7euNmzYoCZNmqhbt27auHGj3wpEYG0vcCnN5VZpnzRuSPJI+jgnv4qrAgAg+OV++YWyly6UkZ8neTxSUZHk8cjIz1P20oXK/fKLCm/zmWeeUdOmTeV0OtWiRQstX77c56psXFycOnTooO7du0uSOnTooKioKL3yyiuSpK+//lrdunVTbGysWrRooXfeece77ZEjR2rcuHG66qqrFBUVpeXLl/vhLPhXha7YdujQQStWrNCQIUNKrIuOjtbq1as1ePBg3XfffbJYLH4rEoGzOa/wtOs9kr7LL1KRYSiUngMAUC5GUZFOrFgmlfV+FcPQiXeXKbzjRRWaltCoUSNt2LBB9erV07JlyzR8+HDt2rVL27dvV6NGjZSRkaGwsDBJJ6cifPPNN97Qe+DAAfXp00evvfaa+vfvr//85z/q27evWrRooRYtWkiSFi9erFWrVum9995Tfn71u7BVoSu2N9xwg/bu3aujR4+Wut7hcGjFihUaM2aMGjZs6JcCEVhZHo/O9BYxj6R83kgGAEC5FezYdnL6wWkYBQUq2PFDhbZ77bXXqn79+rJarRo6dKiaN2+uzZs3l+uxCxcuVEpKigYMGCCbzabOnTtr4MCBWrZsmXfMlVdeqR49eshisSg8PLxCtVWFCl2xHTx4sAYPHnzaMTabTXPnzj2rogLhxIkTmj17tr799luFh4dryJAh6tevX6DLCrg4m1V7i9zynGZMqKQIrtYCAFBunqysk3NqPaf5DWuzyZN1vELbfeONNzRjxgzt3btX0sl8k5GRUa7H7t27V++9955iYmK8y1wul0aMGOH9urpfuKzw29n379+vyMhI1axZs9T1x44dU05Ojho0aHDWxVWluXPnyu126/XXX9eBAwf0yCOPKDExUW3atAl0aQHVJcKhf+cXlbneJqlTuF02gi0AAOVmjY6W3KW9g+UUbres0TXKvc19+/Zp9OjRWrNmjbp27Sqbzab27duX+YFav9ewYUNdd911mj9/ftl1W6v3h9ZWqLovvvhCTZo00e7du8scs3v3bjVt2lSffvrpWRdXVfLz87Vp0yYNHz5cERERatKkiXr27Kk1a9YEurSAaxJqU1tHqEq754FVUpjFoj5RYVVdFgAAQc1xfmtZ7I7TjrE4HHKc36rc28zJyZEkxcfHSzp59faHH8qeylCnTh3t2bPH+/Xw4cP10UcfadWqVXK5XCosLNTmzZu1c+fOctcQaBUKts8995yGDRumTp06lTmmU6dOGjFihGbPnn3WxVWV9PR0Sb6X1xs3bqx9+/YFqqRqw2Kx6KaYCF0aYVeITj5hikNu41Cb7qsVpVhb9f7rDQCA6sYSGqqogYOlsl7xtFgUNWBwhd44dv755+u+++5T165dVadOHX333Xfq0qVLmeMff/xxjR49WjExMZo3b54SExP1wQcf6Pnnn1edOnVUr149PfjggyooOP1c4OqkQlMRNm7cqHnz5p1xXP/+/TV27NhKF1XV8vPzS0yAjoyMVF5ens+yjIwMn3kqcXFxstvtVVJjVbJYLLLZ/neN1iZpcKxTV9bwaHeBS0UylBgaotohwXHv2uJjOfWYzOD3fQp29Ck40KfgQJ+CQ/F9ak+8u0xGwSn3sXU4FDWgcvexnTZtmqZNm1bqut9PSRg3bpzGjRvns+zCCy/U2rVrS3386aYoVBcVCrbHjx9XrVq1zjiuZs2a+u233ypdVFULCwsrEWJzcnJKhN3ly5f7BPsxY8aUeEKYRWmBPVZSQtWX4jfR0dGBLsHvzPiHFX0KDvQpONCn6i+iSzeFd7xIBTt+kCfruKzRNeQ4v5XfP3nsj6JCwbZ27dravXu3LrnkktOO27Nnj2rXrn1WhVWl+vXrSzr5xrjiN72lpqYqKSnJZ9ygQYN06aWXer+Oi4tTVlaW3Gea/B1kIiMjvfN0zMBmsyk6Otp0vaJPwYE+BQf6FBz81afY2Fg/VOM/ltBQhbVtH+gyTKFCwbZHjx6aOXOmhg0bVuZHqBUVFWn27Nnq2bOnXwqsCmFhYeratasWL16sO++8U4cOHdLatWt1//33+4yLi4tTXFycz7LMzExT/dCQTr5UYbZjkiS3222q46JPwYE+BQf6FBzM2if4T4Xe9fPggw9q586d6tu3r7Zv315i/fbt29WvXz9t375dEydO9FuRVaF4SsHIkSP1+OOPa9iwYWrbtm2AqwIAAEB5VeiKbYsWLfTPf/5TI0aMUJs2bVS/fn01aNBAFotFaWlpSk9PV3R0tJYtW+b96LVgERUVFXRhHAAAAP9T4Q9ouPLKK/Xjjz9q3rx5Wr9+vdLT02UYhpo3b67x48dr9OjRQTW/FgAAIJhY+FCkMlU42G7fvl1z585VamqqGjdurPvvv1+XX375uagNAAAgqDmdzkCX8IdS4fvY9urVSy6XS3FxcTp27JheffVVzZ49W+PHjz9XNQIAAABnVKE3jz322GM6//zztXfvXh06dEhHjx7VgAEDNHny5HNVHwAAAFAuFbpi+/3332vOnDnee71GR0fr2WefVePGjX3uAQsAAAApOzv7nGyXKQ6lq9AV24yMDCUmJvosKw6zp37ULAAAAFDVKhRsJd6JBwAAgOqpwndF6NGjh6zWknm4W7duPsstFouOHz9+dtUBAAAA5VShYPvoo4+eqzoAAACAs0KwBQAAMIHk5GTNmTNHffr0qfBjx4wZo+XLlys+Pl4//vjjOaiualR4KgIAAAD8p8gwtC2vQFluj6JtVrUOdyi0Ct/TtHHjRq1atUr79u0L+rstEGwBAAAC5IvsXC3LPKECw5DNIrkNyWGxaHBslLo5I6qkhtTUVCUnJ1cq1LrdblksllLffxUI1aMKAACAP5gvsnO18Fi28gxDHklFhuSRlGcYWngsW19k51Z4m1u2bFHr1q1Vo0YNDRo0SL/99psk6euvv1a3bt0UGxurFi1a6J133pEkvfLKKxozZoy+/vprRUVF6e6775YkzZ8/X82bN1dsbKxSUlL0008/efeRnJysp556Su3bt1dkZKQOHz6srKwsjRs3TomJiapbt65uv/125efnn/U5qiiCLQAAQBUrMgwtyzwho4z1hqRlmSdUZJQ1onTz58/Xe++9p19++UUFBQW66667dODAAfXp00f33nuvMjIyNH/+fI0ePVo7d+7U2LFjNWfOHHXs2FEnTpzQ888/r/Xr12vChAlasGCBDh06pC5duujKK69UUVGRdz+LFi3S8uXLlZWVpfj4eI0aNUoFBQXasWOH/vvf/2rXrl2aMmVK5U9QJRFsAQAAqti2vAIVnCG0FhiGfsgrqNB2b7/9djVu3FhOp1PTpk3T0qVLtXDhQqWkpGjAgAGy2Wzq3LmzBg4cqGXLlpW6jUWLFmnkyJG66KKLZLfb9cgjj+jYsWPavHlzif3Y7XYdPXpUK1eu1MyZMxUdHa2YmBhNnjxZS5YsqVDt/sAcWwAAgCqW5fbIZpE8p8m2Not03O2p0HYbNmzo/X9SUpIKCwuVmpqq9957TzExMd51LpdLI0aMKHUb6enpPndWCAkJUWJiotLT00vdz969e+V2u72fRitJhmHI7XZXqHZ/INgCAABUsWibVe4zzDJwG1INW8VeXE9LS/P5f2hoqBITE3Xddddp/vz55dpG/fr1tW/fvv/V4Xbrl19+Uf369b3LTn2zWMOGDRUSEqLDhw/LbrdXqF5/YyoCAABAFWsd7pDjDLf0clgsahXuqNB2X3rpJaWmpio7O1uTJ0/W0KFD9Ze//EUfffSRVq1aJZfLpcLCQm3evFk7d+4sdRvDhg3TggUL9J///EeFhYWaMmWKYmNj1blz51LH161bV3/+85911113KTMzU4ZhaP/+/fr4448rVLs/EGwBAACqWOj/39KrrGhrkTQ4NqrC97O98cYbddVVVykxMVE2m00vvPCCEhMT9cEHH+j5559XnTp1VK9ePT344IMqKCh9/m6PHj309NNP64YbblDt2rX1+eefa9WqVQoNDS1zvwsWLFBISIjatWunGjVq6IorrvC5k0JVsRhGBd9uBx+ZmZkBmUNyLjmdTmVnZwe6DL+x2WyKjY01Xa/oU3CgT8GBPgUHf/UpLi7OD9WUz5nqrex9bIP9gxTOFebYAgAABEg3Z4QuigrXD3kFOu72qIbNqlZV/MljZkKwBQAACKBQi0XtI8ICXYYpMMcWAAAApkCwBQAAgCkQbAEAAGAKzLEFAAA4R2w2W6BL+EMh2FYjhmHoF5dbv7kNRdssahhik4V3RQIAELQiIsq+ZRf8j2BbTewsKNKyrDwddntkleSRVMtm1SBnuNqElX1DZAAAAJzEHNtqYHtBkV7KzNFht0fSyVArSUfdHs37LUff5hUGrjgAAIAgQbANMI9haMnxXJX18W+GpKVZeXLxAXEAAACnRbANsN2FLh33nD605huGthcUVVFFAAAAwYlgG2BH3R6d6f2Stv8fBwAAgLIRbAMs0mrVmSKrR1KElVYBAACcDmkpwM5zhCj0DHf0skhq4+AGFgAAAKdDsA0wu8WifpFhZTbCKikl0sEVWwAAgDPgMmA10DPSoSLD0Ic5BbLq5J0QLDo5BaFnhEP9osICWyAAAEAQINhWAxaLRX2c4bok0qFv8op03OOR02pVh7BQRdu4UgsAAFAeBNtqJMpq1aWRjkCXAQAAEJQItijh18Ii/VJQpCirRYkhNlksZ3h3GwAAQDVgMQw+0upsFBYWyiyn8Ke8Ai06kql9BUWy6ORc3/gQm66Pj1GHqIhAl1dpFotFdrvdVL2SpJCQELlcrkCX4Tf0KTjQp+BAn07P4eDVUbPiiu1ZysnJkdvtDnQZZ213oUsvHjvhvadu8Y/BIy63Zh44qhtr5KljuD1Q5Z0Vm80mu91uml4Vczqdys7ODnQZfkOfggN9Cg706fQItubFO5MgwzD05vHcMj8owpC0NCtXhSb6qx8AAJgPwRZKc7l1+Awf2esypO/zi6qoIgAAgIoj2EIZLs8Z56RYJWWcIfwCAAAEEsEWirBadKYZWMb/jwMAAKiuCLbQn+whCjtDZvVIaucIrZJ6AAAAKoNgC4VYLLoyKlxlZVubpEsi7HwKGgAAqNZIKpAkdY+w68ooh6ySQiwnw2zxvNuLwu0a5AwPYHUAAABnxn1sIenkzbyviApXlwiHfvBYdSA3T06rRR3C7arJlVoAABAECLbw4bRa1buGU9k8MwAAQJDhUhwAAABMgWALAAAAUyDYAgAAwBQItgAAADAFgi0AAABMgWALAAAAUyDYAgAAwBQItgAAADAFgi0AAABMgWALAAAAUyDYAgAAwBQItgAAADAFgi0AAABMgWALAAAAUyDYAgAAwBQItgAAADAFgi0AAABMgWALAAAAUyDYAgAAwBQItgAAADAFgi0AAABMgWALAAAAUyDYAgAAwBQItgAAADAFgi0AAABMgWALAAAAUyDYAgAAwBQItgAAADAFgi0AAABMISTQBVTUQw89pG3btuntt9+W3W6XJJ04cUKzZ8/Wt99+q/DwcA0ZMkT9+vXzPuaHH37QnDlzdPDgQSUnJ+vOO+9Uw4YNvesXLVqkjz76SB6PR926ddPYsWMVEhJ0pwYAAOAPLaiu2K5du7bU5XPnzpXb7dbrr7+uhx9+WIsXL9b3338vScrKytK0adN07bXXasmSJercubOmTp0qt9stSVq9erU+//xzzZgxQ3PmzNHPP/+sf/7zn1V2TAAAAPCPoAm2WVlZWrZsmUaOHOmzPD8/X5s2bdLw4cMVERGhJk2aqGfPnlqzZo0k6auvvlJCQoIuu+wyhYaG6pprrlFeXp5++OEHSdKaNWs0YMAA1alTRzVq1NCQIUO8jwUAAEDwCJrX2//xj3/oqquuUnR0tM/y9PR0SfKZWtC4cWO9++67kqS0tDQ1atTIu85msykpKUn79u1T27ZtlZaWpuTkZO/6Ro0aKSMjQzk5OYqMjPTZV0ZGhjIyMrxfx8XFeadDmInFYpHNZgt0GX5TfCxmOiaJPgUL+hQc6FNwMFuf4H8BD7bFUwLKYrPZtG3bNu3fv1933nmnjhw54rM+Pz9f4eHhPssiIyOVl5cnScrLy1NUVFSZ6/Pz830CbPH/8/LySgTb5cuXa968ed6vx4wZo3HjxpXnMIOOGQP77/8oMgP6FBzoU3CgT8HBjH2C/wQ82D788MPeaQG/FxMTo9dee01z5szR3XffLau15MyJsLAwb0gtlpOT4w274eHhys3N9Vmfm5vrXR8WFuazvvj/vw/LkjRo0CBdeuml3q/j4uKUlZV1xnAebCIjI5WTkxPoMvzGZrMpOjradL2iT8GBPgUH+hQc/NWn2NhYP1SD6ijgwfZvf/vbadcfOnRI6enpeuyxxyRJhmFIkkaNGqV77rlHLVu2lCTt379fDRo0kCSlpqYqKSlJ0skpCp9++ql3ex6PR3v37tW1117rXZ+amqoWLVp4HxsXF1fiaq10MsjGxcX5LMvMzDTVDw3p5Dk22zFJJ18dMNNx0afgQJ+CA30KDmbtE/yn2r95LC4uTq+++qpeeOEFvfDCC3rkkUckSdOnT1fr1q0VFhamrl27avHixcrNzVVqaqrWrl2rXr16SZIuvvhipaena8OGDSoqKtKKFSsUHh6uVq1aSZJ69eqllStX6vDhw8rKytJbb72llJSUgB0vAAAAKifgV2zPxGaz+VwlLf5LrVatWt55NuPGjdOsWbM0cuRIRUREaNiwYWrbtq2kk/OLJk2apDlz5mjmzJlKTk7W5MmTvZPPe/furSNHjmjChAlyu93q3r27hgwZUsVHCQAAgLNlMYpf20elmHEqgtPpVHZ2dqDL8BubzabY2FjT9Yo+BQf6FBzoU3DwV59+P60Q5lHtpyIAAAAA5UGwBQAAgCkQbAEAAGAKBFsAAACYAsEWAAAApkCwBQAAgCkQbAEAAGAKBFsAAACYAsEWAAAApkCwBQAAgCkQbAEAAGAKBFsAAACYAsEWAAAApkCwBQAAgCkQbAEAAGAKBFsAAACYAsEWAAAApkCwBQAAgCkQbAEAAGAKBFsAAACYAsEWAAAApkCwBQAAgCkQbAEAAGAKBFsAAACYAsEWAAAApkCwBQAAgCkQbAEAAGAKBFsAAACYAsEWAAAApkCwBQAAgCkQbAEAAGAKBFsAAACYAsEWAAAApkCwBQAAgCkQbAEAAGAKBFsAAACYAsEWAAAApkCwBQAAgClYDMMwAl1EMCssLJTZTmFISIhcLlegy/Abi8Uiu91uul7Rp+BAn4IDfQoO/uqTw+HwQzWojkICXUCwy8nJkdvtDnQZfuV0OpWdnR3oMvzGZrPJbrebrlf0KTjQp+BAn4KDv/pEsDUvpiIAAADAFAi2AAAAMAWCLQAAAEyBYAsAAABTINgCAADAFAi2AAAAMAWCLQAAAEyBYAsAAABTINgCAADAFAi2AAAAMAWCLQAAAEyBYAsAAABTINgCAADAFAi2AAAAMAWCLQAAAEyBYAsAAABTINgCAADAFAi2AAAAMAWCLQAAAEyBYAsAAABTINgCAADAFAi2AAAAMAWCLQAAAEyBYAsAAABTINgCAADAFAi2AAAAMAWCLQAAAEyBYAsAAABTCAl0AQAAAJJkGIZce3ap8D+bZWRnyRpbU/bOXRTSICnQpSFIEGwBAEDAGQUFOvH6XLn37JIsFsnjkaw2Ff5rk0LbX6iI60YEukQEAaYiAACAgMtdulDu1D2SYZwMtZLkcUuSirZuUd6HKwNYHYIFwRYAAASU+2iGirZ9J7ndpQ/wuFW4aYM8eXlVWxiCDsEWAAAElGvHD1LIGWZHejwq+HFn1RSEoEWwBQAAAWUUFUmynH6Q1SqjqLBK6kHwItgCAP6vvXuPaer8/wD+bsutrYJosYhgxWu8flX8Q5ibXLZMNzVzOpwWjRpRcepM3PZz2TA0cUwAABD0SURBVFzMpsYszunAecFMvOBkQWcczhkLTAXZkolxMHSZ2hYHQ8ZQEVquPb8/GGdWJnVyqef4fiUm5Xx68Hn6yfOcD+3T5xC5lapPENDU2PaTGhvhERTcNQ0iyWJhS0RERG7lMXQYFN19H/4EpRKqEAO8+rKwpbaxsCUiIiK3UiiV0BoXAipV81Zf91MqAS8vaGbHuadxJCksbImIiMjtPAYOQrdVb8Fj2AiI622VSnj+bxy6r/4/qAL7uLV9JA2SuEFDQ0MDkpOTkZeXh8bGRgwePBjLli1DUFCQGN+zZw9ycnKgUqkwZcoUGI1G8Xyr1YrExERYLBYEBgYiISEBI0aMEOMZGRlIT0+H3W5HWFgYVqxYAY1G0+X9JCIiepp59A1Bt0XLINTaIdjsUGi1UHh7u7tZJCGSeMf2+PHjuHLlCrZv346DBw8iJCQEW7duFeNpaWmwWCzYvXs3PvnkE5w9exYmkwkA0NjYiA0bNiA8PBxffvklZs6ciY0bN6K6uhoAcOnSJRw5cgQffPAB9u3bh4aGBuzevdst/SQiIiJA4aOGsmdPFrX0n0misL116xbGjRuHnj17wsPDA1FRUSguLhbjJpMJs2fPhq+vL/R6PaZPny4WtgUFBairq8OMGTPg6emJqKgo6PV6XLhwAQCQlZWFmJgYDBgwABqNBkajETk5Oairq3NLX4mIiIjo8UiisH3hhRdQVFSEiooK1NfXw2QyISwsDABQXV2NyspKhIaGis8PDQ2F1WoFABQXF8NgMECpVDrFWwpjq9XqdK7BYIDD4UBpaWlXdI2IiIiIOojb19g2Pez2eX9TqVQIDg5G7969sWjRIiiVSgQGBuKjjz4CANj/vr3e/WtitVqteNxut0Or1Tr9Tq1WC5vNBgCora11iisUCmg0GvH8+1VUVKCiokL8WafTwcvL6790VxIUCgVUKpW7m9FhWvoipz4BzJNUME/SwDxJg9zyRB3P7YXtunXrUFhY+K+xHj164MCBA/j8889ht9tx8OBBaLVamEwmvPvuu9ixYwfUajUAwGaz/etjtVotFrEt7o/7+Pi0Gb/f0aNHkZycLP4cHx+PpUuXPmbPn2xyLNh9fdvYI1GimCdpYJ6kgXmSBjnmiTqO2wvbTZs2uXyO1WrFnDlz4OfnBwCYPHkyvvjiCxQXF2Pw4MHo2bMnLBYLevXqBQC4ceMGDAYDAKBfv344duwYHA6HuBzBbDZj8uTJAJqXHpjNZkyaNEn8v5RKpbjjwv1mzpwpPg9ofse2qqrK5bvOUqPValFTU+PuZnQYlUoFX19f2eWKeZIG5kkamCdp6Kg8+fv7d0Br6Enk9sL2UQwZMgTZ2dkYPXo0NBoNMjMzAQB9+jTvaRcTE4O0tDQMHjwYdrsdJ06cQGxsLABg1KhR8PT0xPHjxzFt2jRcuHABZWVlCA8PBwBER0dj69atmDRpEvR6PVJTUzFx4kR4/8s3MXU6HXQ6ndOx27dvy2rSAABBEGTXJ6B52Yuc+sU8SQPzJA3MkzTINU/UcRSCIAjuboQr9+7dw969e5Gfn4+GhgYEBQVh3rx5GDt2LIB/9rE9f/68uI9tXNw/dyixWCxISkqCxWKBXq9HQkICRo4cKcZb9rG12WwICwvDypUrH3kfWzkWtt27d8e9e/fc3YwOo1Kp4O/vL7tcMU/SwDxJA/MkDR2VpwffpCL5kERh+yST26QBcIKXCuZJGpgnaWCepIGFLbkiie2+iIiIiIhcYWFLRERERLLApQiPqaKiAkePHsXMmTP5kcYTjrmSBuZJGpgnaWCe6GnFd2wfU0VFBZKTk51u2EBPJuZKGpgnaWCepIF5oqcVC1siIiIikgUWtkREREQkC6r169evd3cjpEqtVmP8+PGPvOctuQ9zJQ3MkzQwT9LAPNHTiF8eIyIiIiJZ4FIEIiIiIpIFFrZEREREJAssbImIiIhIFjzc3YAnwc8//4y0tDRcv34dXl5eOHDggFO8uroaO3bsQH5+PtRqNWJjY/HSSy+J8cLCQuzatQtlZWXo378/Vq1ahX79+onxQ4cO4dSpU3A4HHj22WexZMkSeHg0v/R//vknEhMTceXKFfj7+2PBggWIiIjomo5LXGZmJhITE+Hl5SUeW758OSIjIwEADQ0N2LNnD3JycqBSqTBlyhQYjUbxuVarFYmJibBYLAgMDERCQgJGjBghxjMyMpCeng673Y6wsDCsWLGCX8LoJK7GGHW8bdu24dy5c+JcBAA7duxAQEAAANdzU25uLlJSUnDnzh0MGzYMb775Jnr16gXA9dijtmVkZCArKwsWiwXh4eF4++23xVh75q32XsuIJEEg4ddffxWysrKE06dPC/PmzWsV37Jli7Bx40ahpqZGuHbtmjB37lzh8uXLgiAIwt27d4XXX39dyM7OFurr64WvvvpKiI+PFxobGwVBEITTp08L8fHxQllZmXDnzh1hzZo1Qmpqqvi733nnHWHXrl1CbW2tkJ+fL8TGxgq///5713Rc4kwmk7BmzZqHxg8ePCi89dZbwt27d4WysjIhPj5eOHPmjCAIgtDQ0CAsXrxYSE9PF+rr64WsrCxhzpw5wr179wRBEIT8/HzBaDQK169fF2pqaoQNGzYIW7du7ZJ+PY3aGmPUOT799FMhJSXlofG25qabN28KsbGxwqVLl4Ta2lph586dwtq1a8Vz2xp75Fpubq6Ql5cn7Ny5U/j444/F4+2dt9pzLSOSCi5FADBkyBBERUWhT58+rWK1tbXIzc1FXFwcNBoNBg4ciOjoaJhMJgBAXl4egoKCEBkZCU9PT7z66quw2+0oLCwEAJhMJrzyyivQ6/Xw8/NDbGyseG5paSl+++03GI1GeHt7Y+zYsRgzZgyys7O7rvMyZjKZMHv2bPj6+kKv12P69Onia19QUIC6ujrMmDEDnp6eiIqKgl6vx4ULFwAAWVlZiImJwYABA6DRaGA0GpGTk4O6ujp3dkmWXI0x6nqu5qbs7GyMGzcOY8aMgbe3N4xGI65evYo//vgDQNtjj1yLiIjAhAkT4Ovr63S8PfNWe69lRFLBwtaFkpISAHD6OGbAgAGwWq0AgOLiYoSGhooxlUoFg8HgFO/fv78YDw0NRUVFBWpqamC1WhEQEIBu3bo5xYuLizuzS7JisVgQFxeHJUuWICUlRSw8q6urUVlZ6ZSb0NBQp7wYDAYolUqneMtrb7Vanc41GAxwOBwoLS3tim49VVyNMeo8p0+fxty5c7Fq1SqcOXNGPO5qbnpwfHTv3h0BAQGwWq0uxx49vvbMW+29lhFJhezX2DY1NbUZV6lUbcZra2uhVqudjmm1WtjtdgCA3W53mvwfjNfW1kKr1TrFWs57MAYA3bp1E899mj1K3kaMGIGkpCT07t0bpaWl2LZtG1JSUrB06VLxNbx/TeyDeXvwtddqtbDZbABa502hUECj0TA3ncDVGKPOMW3aNCxatAharRZFRUXYvHkztFotIiIiXM5NtbW1rdabt+TM1dijx9eeeUsQhHZdy4ikQvaF7bp16x76UUqPHj1afVHsQT4+Pq0Gdk1NjThBqNVqcVJpYbPZxLiPj49TvOWxWq1uFXvwdz/NHiVvgYGB4rHg4GDMnz8fW7ZswdKlS8XX8P5c3P/4v+btwTh1HFdjjDrHwIEDxcejRo3Cyy+/jNzcXERERLicm9qKuxp79PjaM285HI52XcuIpEL2he2mTZvadX7fvn0BADdv3kRISAgAwGw2w2AwAGj+WOf+j/AcDgcsFgtmzZolxs1mM4YNGyaeq9PpoNVqYTAYUF5ejpqaGvGvbLPZzG+h4vHyplQqIfx9I71u3bqhZ8+esFgs4je1b9y44ZS3Y8eOweFwiB/rmc1mTJ48GUDzR3hmsxmTJk0C0PwRn1KpRFBQULv7Rs5cjTHqGgqFQhw/ruYmg8EAi8UinltdXY2KigoYDAaXY48eX3vmrZbcPu61jEgquMYWzQO4vr4ejY2NAID6+no0NDQAaP4L+JlnnkFqaipsNhvMZjMyMzMRExMDAAgPD0dJSQnOnj2LhoYGfP3111Cr1Rg5ciQAICYmBidOnEB5eTmqqqqQlpaG559/HgAQFBSEQYMG4fDhw6irq8Ply5dx6dIlREVFueFVkJ6LFy+isrISAFBWVob9+/cjPDxcjMfExCAtLQ1VVVW4desWTpw4Ib72o0aNgqenJ44fP46GhgacPXsWZWVl4vnR0dHIzMyE2WyGzWZDamoqJk6cCG9v767vqMy5GmPUOXJycmCz2eBwOFBUVISTJ09iwoQJAFzPTZGRkbh48SIuX76Muro6pKamYujQoeIXcNsae+RaU1MT6uvr4XA4nK5P7Zm32nstI5IKhdDyZ9xTrKCgAO+9957Tsd69e2Pv3r0Amt+NSEpKQn5+PjQaTau9/woKCrBr1y7cunWr1d5/giAgNTUVp06dQlNTE5577rlW+9h+9tln4l6RCxcu5D62j2jfvn3Izs6G3W6Hr68vIiIiYDQa4ePjA+CfvTTPnz8v7qUZFxcnnm+xWJCUlASLxQK9Xo+EhASnSbxlP0ibzYawsDCsXLmS+9h2EldjjDre2rVrYbVa4XA4oNPpMHXqVEyZMkWMu5qbcnJysH//fty+fRvDhw//131sHzb2qG2HDx/GkSNHnI5FR0dj9erV7Zq32nMtI5IKFrZEREREJAtcikBEREREssDCloiIiIhkgYUtEREREckCC1siIiIikgUWtkREREQkCyxsiYiIiEgWWNgSERERkSywsCUiIiIiWWBhS0SSt379eigUCvFfQEAAYmJicP78eafnFRUVYf78+QgODoaXlxcCAgIwY8YMnDt3TnzOtWvXsGzZMowZMwYeHh68pSgRkYSwsCUiWVCr1cjLy0NeXh527tyJv/76CzExMSgoKADQfKvRsLAwFBQU4MMPP4TJZMKePXvg5+eH6OhoWK1WAMAvv/yCkydPYtCgQRg+fLg7u0RERP8Rb6lLRJK3fv16bNmyBdXV1eKx4uJi9O/fH8uXL8e6deswdOhQjB07Ft999x28vb2dzv/hhx8waNAg6HQ6OBwOKJXNf/MvWLAAP/30EwoLC7u0P0RE9Hg83N0AIqLO0K9fP+h0OpjNZiQnJ+Pu3bvYvn17q6IWACZMmCA+bilqiYhIejiDE5EsVVVVobKyEkFBQfj+++/Rt29fjB492t3NIiKiTsTClohko7GxEY2NjbBYLFi4cCGampowa9YslJSUICQkxN3NIyKiTsalCEQkCzU1NfD09BR/9vf3R1JSEl588UUIggCFQuHG1hERUVdgYUtEsqBWq3Hu3DkoFArodDqEhISI62WDg4Nx9epVN7eQiIg6G5ciEJEsKJVKjB8/HmFhYTAYDE5fAouKikJJSYm49RcREckTC1sikr3FixfDz88Pq1evRn19fav4jz/+iIqKCje0jIiIOhKXIhCR7On1ehw6dAizZs1CREQE3njjDQwcOBCVlZXIyMjA/v37ce3aNeh0OthsNnz77bcAAKvViqqqKqSnpwMAJk2ahICAAHd2hYiI2sDCloieClOnTsXFixexefNmvP/++ygvL0ePHj0QHh6Ob775BgaDAQBQXl6O1157zenclp+zs7MRGRnZ1U0nIqJHxDuPEREREZEscI0tEREREckCC1siIiIikgUWtkREREQkCyxsiYiIiEgWWNgSERERkSywsCUiIiIiWWBhS0RERESywMKWiIiIiGSBhS0RERERyQILWyIiIiKSBRa2RERERCQL/w/B4pGTOcOS1QAAAABJRU5ErkJggg==\n",
            "text/plain": [
              "<Figure size 640x480 with 1 Axes>"
            ]
          },
          "metadata": {
            "tags": []
          }
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "<ggplot: (286316082)>"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 24
        }
      ]
    },
    {
      "metadata": {
        "colab_type": "code",
        "id": "tLK4S3Ez9ihL",
        "outputId": "2ec2e880-7886-43a6-8460-c1ec29a76a74",
        "scrolled": false,
        "colab": {}
      },
      "cell_type": "code",
      "source": [
        "# graph t-SNE!\n",
        "chart = ggplot(dataframe, aes(x='tSNE1', y='tSNE2', color='label') ) \\\n",
        "        + geom_point(size=3,alpha=1) \\\n",
        "        + ggtitle(\"tSNE dimensions colored by digit\")\n",
        "chart"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAqcAAAHICAYAAABzr0MWAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAAPYQAAD2EBqD+naQAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAIABJREFUeJzs3Xd4FPXe/vF7drO76SGQhBapCoIgIFKkg4h6VBA5KApHQEHwyIOKgtgoUmyoyEEEsYAKPM8PkapHlKpYUMSGKKBUEcFAIL1t5vcHJ3tcUtgkm+wkeb+uKxfszHdnPrv5kNxM+a5hmqYpAAAAwAJsgS4AAAAAyEM4BQAAgGUQTgEAAGAZhFMAAABYBuEUAAAAlkE4BQAAgGUQTgEAAGAZhFMAAABYBuEUAAAAlkE4BfygR48eatCgwXmXVRbDhg2TYRiBLiOgDMPQsGHDAl1GkQ4ePCjDMDRlypTzjl20aJEMw9CWLVvKvK7z2bJliwzD0KJFi4pcVlxTpkyRYRg6ePBgqWsEUHYIpyhzBw8e1JQpU/Ttt98WuD4rK0sLFixQx44dFRcXp+DgYF1wwQXq1auXJk2apMzMTM/YvF+ghmFo5cqVBW7PMAz17t3ba1mPHj08zyvoy+ohA0DZ+PbbbzVlyhQCK2AhQYEuAJXfwYMHNXXqVDVo0ECtW7f2Wud2u9WnTx9t3bpV3bp10/jx4xUdHa3Dhw/ryy+/1MyZMzV27Fi5XK582504caJuuOEGBQX51sY2m02LFy8ucF3jxo2L/8LO48MPP5Rpmn7frhUsXLhQ8+fPD3QZqEK6deum9PR0ORyOEm/jscce08SJE71+nnz77beaOnVqpT7TAVQ0hFME1Jo1a7R161bddNNNWrFiRb71x48fV1RUVL7l7dq101dffaVXXnlF//znP33al2EYGjJkSKlr9pXT6Sy3fZU3h8NRqpCAgmVkZCgoKMjn/3BVJTabTcHBwaXaBu8tUDFwWh9lasqUKerZs6ckafjw4Z7T6D169JAk7dmzR5I8Y85Vs2bNAkPQ6NGj1ahRI02dOlUpKSllU3wBfvnlF/Xv319RUVGKiIhQnz599N133xU4tqjrUA8fPqyBAwcqOjpaUVFRGjBggE6cOCFJev3119WiRQsFBwerYcOGeu211wrc/ubNm3XttdcqOjpaLpdLzZo109NPPy23213gPn///Xfddtttql69ukJCQtS9e3d9/fXXXmNN09S//vUvtWnTRlFRUQoLC1ODBg1066236vjx455xhV1z+vPPP2vQoEGqWbOmXC6XGjVqpAcffFBJSUle4/Iuz9i0aZNmzZqliy66yDP+hRdeyLfdL774QjfccIPq1Kkjl8uluLg4de/eXWvWrCnwvTlXSkqKpkyZohYtWigkJETR0dFq166d5s6d6zXu9OnTGjdunBo2bCiXy6WaNWvq1ltv1b59+3zajyQtWbJEHTp0UHh4uMLCwtSxY0f97//+b75xed+XQ4cOadCgQYqJiVFISIh+++03z5gVK1aoe/fuioyMVEhIiNq0aaNXX321wP0uW7ZMrVq1UnBwsOrWratx48YpLS3N57rz5OTk6IknnvC8BxdffLHmzZvnNeaee+6RYRjavXt3vudnZGSoevXq6tKli0/7mzNnjpo2bSqXy6WGDRtq2rRpysnJyTeusGtOk5KSNGbMGNWqVUshISG67LLLtHz58gKvLz132bBhwzR8+HBJZ38GcZkPYA38FxJl6qabblJ2drZmzpypu+66S127dpV0NnRKUqNGjSRJy5cv1+DBgxUdHe3Tdp1Op2bMmKFbb71VzzzzjJ544gmfnpeQkFDg8sjIyPMe6Txy5Ig6deqk06dPa/To0WrWrJk+//xz9ejRQzVq1PBp/5KUmpqqHj16qHPnznryySf1888/a+7cufrjjz/Uv39/vfTSSxo5cqQiIiK0cOFCjRgxQhdffLE6d+7s2cbrr7+uESNGqE2bNpo4caKqVaumTz/9VA8//LC++eabfGEoNTVV3bp10+WXX67p06fr+PHjeuGFF3TNNddo//79ioiIkCTNnDlTjz32mK677jqNGDFCDodDR44c0b///W/98ccfnu9bQb799lt169ZNbrdbd999txo1aqRt27bpueee04YNG/TZZ58pNDTU6zmPPPKIkpOTdccddyg8PFxvvvmmxo0bp9q1a2vQoEGSpL1796p3796KjY3VPffco9q1a+vPP//Ujh079MUXX6hv375Fvt9nzpxR165d9cMPP6hfv3668847ZbfbtWvXLr377rsaM2aMJCk5OVmdO3fW7t27NXjwYHXq1Em//vqr5s2bpw8++ECffvqpmjdvXuS+Jk2apGnTpunSSy/V448/Lkl6++23deutt2r//v165JFHvManpKSoa9euuvzyyzV16lQlJycrPDxc0tkgNXXqVPXs2VOTJ09WSEiI1q9fr5EjR+qXX37RU0895dnO/Pnzdffdd6tJkyaaPHmyHA6HlixZoo8//rjIegvy0EMPKSkpSSNHjpTL5dLSpUt1zz336MSJE54bq0aPHq158+bp1Vdf1fPPP+/1/BUrVigxMVEjR448774mTpyop59+WpdffrmefPJJpaen6/XXX9fq1at9qjUnJ0fXXHONPv/8cw0cOFA9evTQkSNHdOedd6pp06bnff6oUaPkcrn0yiuv6JFHHlGzZs0klc1lPgCKwQTK2ObNm01J5htvvJFvXWZmpnnZZZeZkszQ0FCzd+/e5qOPPmquXbvWTE1NzTf+jTfeMCWZb731lpmbm2u2a9fODAsLM3///XfPGEnmlVde6fW87t27m5IK/Vq2bNl5X8c//vEPU5K5cuVKr+VPPvmkKcmsX79+vn0WtEyS+eSTT3otv//++01JZt26dc3Tp097lv/xxx+my+UyBw0a5Fl27NgxMzg42LzxxhvN3Nxcr+3MmjXLlGRu2bIl3z5nzpzpNXbZsmWmJHPBggWeZW3atDGbNWt23vdi6NCh5rk/Prp27WoahmF+/vnnXsunTp1qSjKnTZvmWZb3fbz00kvNjIwMz/KUlBSzRo0a5hVXXOFZ9uKLL5qSzO3bt5+3roLcc889piRz9uzZ+da53W7P3x9//HFTkvn00097jdmyZUuBPSXJHDp0qOfx3r17TZvNZl522WVmWlqa12tq0aKFabfbzQMHDniW531fHnrooXx17dy50zQMwxw7dmy+dWPGjDFtNpv566+/mqZpmqdPnzbDw8PNevXqefVOWlqa2bp1a1OSOXny5ILfnL/I+57Ex8ebiYmJnuXp6elm27Zt89XfqVMnMyYmxszMzPTaTvfu3c2oqKgC//3+1b59+0ybzWa2b9/eqwdOnjxp1q5dO9/PjIJ+jrzyyiumJHP8+PFe296xY4dpGIYpyavmyZMn51uW97o3b95cZL0Ayg+n9RFQTqdTW7du1cyZM3XhhRdq48aNmjFjhm644QbVqlVLzz33XKHPNQxDzz77rFJTUzV58uTz7stms+mjjz4q8Kuwywry5ObmatWqVWrevLluvPFGr3X33Xef52iXL2w2m+677z6vZXlHlIcOHep1jW3NmjXVtGlTr9PK77zzjjIyMjRixAidPHlSCQkJnq/rr79ekrR+/fp8+7z//vu9ll111VWSzh6ZzBMdHa3ffvtNW7du9fn1SNKff/6pTz75RFdffbU6duzote7BBx9UWFhYgdcUjxkzxuvmlLCwMF1xxRX5apKklStXKj09vVh15ebmaunSpWrUqJH+53/+J996m+2/PwJXrFihqKgo3XvvvV5junfvrp49e2rTpk1KTEwsdF+rVq1Sbm6uJkyYoJCQEK/XNH78eLnd7gKPCD700EP5li1ZskSmaerOO+/0+v4mJCSob9++ys3N1YYNGySdvfEuJSVFY8aM8eqdkJAQPfjgg0W8OwX75z//qWrVqnkeBwcHa9y4cfnqHz16tBISErxmzdi7d6+2bt2qIUOG5DtKfq6VK1cqNzdXDz74oFcPVK9eXffcc49Ptebt+9z3sG3bturTp49P2wBgPYRTBFx4eLgefvhhfffdd0pKStInn3yihx56yPOLa9myZYU+t3v37rr++uv1+uuvF3j921/lTTFV0FdRp6sl6cSJE0pOTi7wtG5wcHCxTgPWqVMn340deQEs7zKHc9edPHnS8/inn36SJF1//fWKjY31+rr44oslyev60ML2mXcpwl+3/dRTTykiIkI9evRQrVq1NHDgQM2fP19nzpwp8jXt379fknTJJZfkWxcaGqrGjRvr119/zbeuoNdbo0YNr5oGDRqk6667Tk899ZSio6PVrVs3PfbYY9q1a1eRNUlnL+NITExUq1atvIJoYa/hwgsvLHBmiJYtW8o0TR04cKDI50sFvwctW7aUpHzvQWxsbIGXsuR9j1u1apXve5wXuvK+x3nbLKg3C6rlfAraTt6yX375xbNs4MCBqlGjhtc1sHl/9+WUvj/q3r9/v2JiYgq8rCbv3wKAiodrTmEp4eHh6tKli7p06aKePXvqmmuu0WuvvaZbb7210Oc8/fTT+ve//62HHnpIa9euLcdq/8ssxpRRdru92Ov+uv3c3FxJZ4NA/fr1Cxxfp04dn/f51223a9dO+/bt00cffaTNmzdr69ateueddzRp0iR98sknhV7Hl7eN4k7MX1RdeRwOh9atW6edO3dq/fr12rZtm1544QXNnDlTzz77rB544AGfXpsvSvPBAiV5Dwo7upj3PV63bl2BYVkqONgXVpO//PW1BQcHa+jQoXrhhRd04MABxcfHa/HixWrfvr1atWpVqv0Up+6q/mEQQGVEOEWZK+kvj06dOkmSjh49WuS45s2ba/jw4Xr11VeLfTraV3FxcYqIiCj07uT9+/cX66ao0mjSpImks0dUz/2wAX8IDQ1Vv3791K9fP0lnA9INN9ygZ555ptCZA/KOHBd0NDM9Pd1zVLI0LrvsMl122WWSpFOnTqljx4569NFHde+99xY6PVDekcnvvvtOubm5RR49bdy4sfbt26fMzMx8gXDXrl0yDEMNGzYs9Pl5r2/Xrl35jvzlvS++HmFv0qSJPvjgA9WuXdvzmouqW5J2796t6667zmvd+c4mFGT37t2e7/252zm3/lGjRun555/Xa6+9ptatW+vEiROaMWOGT/v5a93nvl++1t2oUSPt2bNHJ0+ezPfvL+/o8/kQbgHr4bQ+ylze9ZinTp3Kt+67774rNHzmXU92vjukJemJJ55QaGioJkyYUIpKC2ez2dSvXz/t3r1bq1at8lo3e/bscp3O6uabb1ZwcLCmTJlS4H7T09OVnJxcom3/+eef+Za1bdtWUsHfvzyxsbHq2rWr1q9fry+//NJr3XPPPaeUlBQNGDCgRDUVNMNC9erV1bBhQ2VmZio1NbXQ59psNt12223av3+//vWvf+Vbn3eEUjo7s8SZM2fyjdu2bZs2bdqkXr16FTmbxI033iibzaZZs2YpIyPDszwtLU3PPvus7HZ7vtBXmNtvv12S9PDDDys7Ozvf+jNnzng+Oa1Pnz4KCwvT3LlzvS6/yMjI0KxZs3za31/NmzdPp0+f9jzOzMzU888/L7vdnm9mhCZNmqhnz55atGiR5s+fr4iICM8sC+dz4403yjAMzZo1S1lZWZ7lp06d0ksvveTzNqSzZ0/+6uuvv9ZHH33k0zaK+vkEIDA4cooy17x5c0VERGjevHkKDQ1VtWrVFBcXp169emnz5s0aP368evfurS5duqhOnTpKTk7WF198oXfeeUeRkZE+3exUu3ZtjRs3TtOnTy90jGmaevvttwtcFxsbq6uvvrrIfUyfPl0ffPCBbrnlFo0ePVoXX3yxvvjiC61Zs0aNGzcucG7GslC3bl0tWLBAd9xxh5o2baqhQ4eqUaNGOnXqlH766SetXLlSq1at8swlWxzNmjVThw4d1KFDB9WtW1cnT570zEk6dOjQIp87Z84cdevWTb169fKaSmrp0qVq1aqVxo0bV6LXm/e+X3/99WrUqJFsNpu2bNmiDz/80DPn7Pmev2XLFt13333avHmzevToIYfDoR9//FF79uzRxo0bJUnjx4/XihUrNH78eH333XdeU0lFRUVpzpw5Re7nwgsv1KOPPqpp06apY8eOGjx4sKfnfvjhB82YMcPnTyBq27atpk+frscee0wtWrTQrbfeqvj4eJ04cULff/+91qxZo927d6tBgwaKiorS008/rTFjxqh9+/YaPny4HA6H3n77bZ8umzhXXFyc2rVrpzvuuENOp1NLly7Vzp079dhjjxV45Hj06NG65ZZbdPToUd11110+3xx40UUX6YEHHtCsWbPUuXNn3XrrrcrIyNBrr72mOnXq6NixY+fdxvDhw/Xaa6/p2Wef1cGDBz1TSc2bN09t27bVjh07zntktF27drLZbJoxY4YSExMVFhamhg0bqkOHDj69DgBlIDCTBKCqee+998w2bdqYLpfLlGR2797dNE3TPHjwoDlz5kzzyiuvNC+44ALT5XKZwcHBZpMmTcy7777bM11Onr9OJXWupKQkMy4urkRTSXXo0MGn17Fnzx6zX79+ZkREhBkeHm5eddVV5jfffFPotFG+LDPNoqfbKuw5X3zxhfn3v//drFmzpulwOMyaNWuaV1xxhTlt2jTz5MmT532+aeafDunJJ580u3fvbsbFxZkOh8OsXbu2ee2115obNmzwel5BU0mZpmnu3r3bvPnmm82YmBjT4XCY9evXN8eNG+c1xZFpFj19z7nb3rx5s3nLLbeYDRo0MENCQszIyEizVatW5jPPPGOmp6cX+LrOdebMGfORRx4xmzRpYjqdTrNatWpm+/btzXnz5nmNO3XqlHnfffeZ9evXNx0OhxkTE2MOGjTI3LNnT75tnvve5XnrrbfM9u3bmyEhIWZISIjZoUMHc+nSpfnGFfV9yfPBBx+Yf/vb38waNWqYDofDrFOnjtmzZ0/zueeey/fa3377bbNly5am0+k0a9eubd5///3mjz/+WOyppD766CNzypQpZv369U2n02k2adLEnDNnTqHPy8rKMmvWrGlKMnfs2HHe/fxVbm6u+cILL5gXXnih6XA4zAYNGphPPPGE+dFHH/k0lZRpmmZiYqJ59913m3FxcabL5TIvu+wy89133zXHjRtnSjKPHz/uGVvQVFKmaZqLFi0ymzVrZjocjkK/rwDKj2GalfTDvwEAZc7tdqt+/fqKi4vTzp07A12Ox3XXXaetW7cqKSnpvDM1ALAW/sUCAEps+fLlOnr0qO6+++6A7L+gj2jdsWOHPvjgA/Xu3ZtgClRAHDkFABTb2rVrdfjwYU2bNk3BwcHas2dPodNelaWhQ4cqKSlJnTt3VkREhHbt2qVXX31Vdrtdn3/+uWeOWQAVB+EUAFBsDRo00O+//67WrVtr/vz5553yqqwsWbJEL730kvbs2aOkpCRVr15d3bp10+TJk9WiRYuA1ASgdAinAAAAsAwuxgEAAIBlEE4BAABgGYRTAAAAWAbhFAAAAJZBOAUAAIBlEE4BAABgGUGBLiDQEhISAl2CZRiGoZCQEKWnp4sZxkrG6XQqKysr0GVUWPSgf9CHJUcP+kd59GBMTEyZbh+Bw5FTeNhsNoWGhvJxf6UQiE/IqUzoQf+gD0uOHvQPehClwb8+AAAAWAbhFAAAAJZBOAUAAIBlEE4BAABgGYRTAAAAWAbhFAAAAJZBOAUAAIBlEE4BAABgGYRTAAAAWAbhFAAAAJYRFOgCAACVV7Zp6puMbH2Xka1s01R9h12dQl2KtnNsBEDBCKcAgDLxR45bc06lKCXXVO5/lu3NytEHqZm6JTJEXUL5/HUA+fFfVwCA32XmmnrxVIqS/xJMJSlHkinp/5LS9VNmdoCqA2BlhFMAgN99nZGltFxTZhFj1qdklFs9ACqOKn9a3+l0yuXi1JIkGYYhSQoLC5NpFvUrBYUJCgpSREREoMuosOhB/7BCH/6QlC53EetNSb9ku+UIC1OwzTrHSehB/7BCD6LiqvLhNCsrS1lZWYEuwxLsdrucTqdSU1Pldhf1awWFiYiIUHJycqDLqLDoQf+wQh+mZef4NC4xOVnhFgqn9KB/lEcPcmCp8rLOTwQAQKVxgcMu+3nGhBmGQv9zpBIA8hBOAQB+1yXUVeRpfbukHmEu2QinAM5R5U/rAyie9FxT29Iy9Ulapk7nmgoxDHUMcahHWDBzV8KjVpBdN4S7tC4lM99NUXZJdYLs6sVUUgAKQDgF4LOU3Fw9fzJFJ925nqNiqaapLWlZ+iw9W/dXD1cdx/lO5qKquDo8RLFBdn2QkqHfc85OKBViSF1DXbo6LFguG0dNAeRHOAXgs2Vn0r2CaR63pEzT1CunUzU5JsJzxzNwWbBTlwU7lezOVbakKJshO/0BoAicgwPgk9PuXH2fmV3odYS5kk66c7U3y7e7tFG1RNhtqm63EUwBnBfhFIBPDme7db5YYZe0P5vpdwAAJUc4BeATX4938UMFAFAa/B4B4JNGTvt5A2qOpKZOLmUHAJQc4RSAT8JsNrUPcRY6sbpd0gVBdtXnbn0AQCkQTgH47O8RIarvsOf7wWHX2buw74oO4059AECpcP4NgM9cNkP3Vg/XNxnZ+iQtU6fcuQqzGboixKUOIU4FM28lAKCUCKcAisVuGLo8xKnLQ5yBLgUAUAlxWh8AAACWQTgFAACAZRBOAQAAYBmEUwAAAFgG4RQAAACWQTgFAACAZRBOAQAAYBmEUwAAAFgG4RQAAACWQTgFAACAZRBOAQAAYBmEUwAAAFgG4RQAAACWQTgFAACAZRBOAQAAYBmEUwAAAFgG4RQAAACWQTgFAACAZRBOAQAAYBlBgS4AAABUHn/kuLXqRKK+SUlTrmnqQmeQeoS5VN9B5IBv6BQAAOAXX6dnafGZNNkk5fxn2c6MbH2Vka0bw4PVOzw4kOWhguC0PgAAKLXjOW4tPpOmXP03mEqS+z9/rk7J0J7M7ABUhoqGcAoAAErt47RMGUWsNyRtSM0sr3JQgRFOAQBAqf2YmeM5SlqQXEn7snKKGAGcRTgFAAClZvppDEA4BQAApXaR0y57EesNSQ0cRY0AziKcAgCAUusR6irytL4k9QpzlUstqNgIpwAAoNTiHUH6e0SIDHmHC+M/X91DnbrU5QhMcahQmOcUAAD4RY8wl+Iddm3NyNGPaRkyJdV32NUrLFgtXUEyjKLu5wfOIpwCAAC/udAZpDY1opWcnBzoUlBBcVofAAAAlkE4BQAAgGUQTgEAAGAZhFMAAABYRqW6ISolJUUvvfSSdu7cqZCQEN18883629/+FuiyAAAA4KNKFU4XLFggt9utN954Q8eOHdOkSZMUHx+vSy+9NNClAQAAwAeV5rR+RkaGPv30Uw0ZMkShoaFq3LixevXqpQ0bNgS6NAAAAPio0oTTo0ePSpLq1avnWdaoUSMdOnQoUCUBAACgmCrNaf2MjAyFhIR4LQsLC1N6errXsoSEBCUkJHge22w2xcbGlkuNVme3273+RPEZhsH7Vwr0oH/QhyVHD/oHPYjSqDThNDg4OF8QTU1NzRdYV6xYoYULF3oeDxs2TGPGjCmXGiuKyMjIQJdQoTmdzkCXUOHRg6VHH5YOPVh69CBKqtKE07p160qSjhw5ogsuuECSdODAAdWvX99r3IABA9S9e3fPY5vNpsTExPIr1MLsdrsiIyOVlJQkt9sd6HIqpLCwMKWmpga6jAqLHvQP+rDk6EH/KI8ejI6OLtPtI3AqTTgNDg5W586dtWTJEo0dO1bHjx/Xxo0bNWHCBK9xMTExiomJ8TxOSEjgB9A53G4370kJmabJe+cH9GDp0IelRw+WDj2I0qg04VSSRo0apblz52rYsGEKDQ3V4MGD1apVq0CXBQAAAB9VqnAaHh6uiRMnBroMAAAAlFClmUoKAAAAFR/hFAAAAJZBOAUAAIBlEE4BAABgGYRTAAAAWAbhFAAAAJZBOAUAAIBlEE4BAABgGYRTAAAAWAbhFAAAAJZBOAUAAIBlEE4BAABgGYRTAAAAWAbhFAAAAJZBOAUAAIBlEE4BAABgGYRTAAAAWAbhFAAAAJZBOAUAAIBlEE4BAABgGYRTAAAAWAbhFAAAAJZBOAUAAIBlEE4BAABgGYRTAAAAWAbhFAAAAJZBOAUAAIBlEE4BAABgGYRTAAAAWAbhFAAAAJZBOAUAAIBlEE4BAABgGYRTAAAAWAbhFAAAAJZBOAUAAIBlEE4BAABgGYRTAAAAWAbhFAAAAJZBOAUAAIBlEE4BAABgGYRTAAAAWAbhFAAAAJZBOAUAAIBlEE4BAABgGYRTAAAAWAbhFAAAAJZBOAUAAIBlEE4BAABgGYZpmmagiwikpKQkuVyuQJdhCYZhyOl0KisrS1W8LUosKChIOTk5gS6jwqIH/YM+LDl60D/Kowf53V15BQW6gEDLyspSVlZWoMuwBLvdLqfTqdTUVLnd7kCXUyFFREQoOTk50GVUWPSgf9CHJUcP+kd59CDhtPLitD4AAAAsg3AKAAAAyyCcAgAAwDIIpwAAALAMwikAAAAsg3AKAAAAyyCcAgAAwDIIpwAAALAMwikAAAAsg3AKAAAAyyCcAgAAwDIIpwAAALAMwikAAAAsg3AKAAAAyyCcAgAAwDIIpwAAALAMwikAAAAsg3AKAAAAyyCcAgAAwDIIpwAAALAMwikAAAAsg3AKAAAAyyCcAgAAwDIIpwAAALAMwikAAAAsg3AKAAAAyyCcAgAAwDIIpwAAALAMwikAAAAsg3AKAABwHlOmTFF4eHixntOgQQONGTMmYPuvqAinAAAAsAzCKQAAACyDcAoAAFAMqampGjNmjJo2barQ0FA1aNBAo0eP1pkzZwoc/+yzz6pu3boKDQ1Vv379dOzYMa/1mZmZeuSRR1S/fn25XC41a9ZMS5cuLY+XYklBgS4AAACgIklLS5Pb7daMGTMUGxurI0eOaMaMGerfv782bdrkNXblypWqX7++Xn75ZSUmJmrixIm66aab9Pnnn3vG3Hzzzdq2bZsmT56sZs2a6f06aL6uAAAgAElEQVT339eQIUMUHR2ta6+9trxfXsARTgEAAIohNjZWL7/8sudxTk6OGjZsqC5dumjv3r1q0qSJZ11ycrLef/99VatWTZIUHx+v3r1768MPP1SfPn20efNmrVmzRuvXr1efPn0kSVdddZWOHj2qyZMnV8lwyml9AACAYnrrrbfUpk0bhYeHy+FwqEuXLpKkvXv3eo3r2bOnJ5hK0pVXXqnIyEh98cUXkqQPP/xQ1atXV69evZSTk+P5uvLKK/XNN9/I7XaX34uyCI6cAgAAFMPKlSt1++2366677tKMGTNUo0YNHTt2TP3791dGRobX2Li4uHzPj4uL81x3mpCQoFOnTsnhcBS4r2PHjik+Pt7/L8LCCKcAAADFsHz5crVu3VoLFizwLNu6dWuBY0+cOFHgstq1a0uSqlevrtjYWL3//vsFPr+gcFvZ+TWcZmdn69ixY6pXr54/NwsAAGAZ6enpcjqdXsuWLFlS4NjNmzfrzJkzioqKkiRt3LhRSUlJ6tChgySpd+/eeuaZZ+R0OnXppZeWbeEVhM/XnL788stq3LixQkJCdNlll2nZsmX5xuzcuVMNGzb0a4EAAABWctVVV+nLL7/UE088oQ0bNuiBBx7Qxo0bCxwbERGha6+9VmvWrNGbb76pIUOGqH379rr66qs927rhhht0zTXXaPbs2dq0aZPWrl2rp556SiNGjCjPl2UZPh05Xb58ue655x71799fw4cP18cff6zBgwdrzZo1Wrx4cb7/PQAAAFRWo0aN0v79+zV37lzNmjVLV199tZYuXaqOHTvmG9u/f3/Fx8dr9OjRSkxMVO/evb0uB5Ckd955R0899ZTmzZunQ4cOKSoqSi1atNDw4cPL6yVZimGapnm+Qe3bt1enTp00e/Zsz7LVq1dr+PDhatWqlVatWqWoqCht375dnTp1qlB3liUkJAS6BMuw2+2Kjo5WYmJihfoeWklERISSk5MDXUaFRQ/6B31YcvSgf5RHD8bExJTp9hE4Pp3W37Nnj/r27eu1rF+/ftq2bZt+/fVXdevWLd+nHQAAAADF5VM4DQsLK/B/QM2bN9dnn32m7OxsderUSXv27PF7gQAAAKg6fAqnl156qdatW1fguvj4eG3btk2xsbEaOXKkX4sDAABA1eLTDVH9+/fXjBkzdPLkSdWoUSPf+urVq2vz5s266aab9NFHH/m9SAAAgEApi+tnHQ6HgoOD/b7dysCnI6ejRo3S4cOHCwymecLCwrR+/Xrl5ub6rTgAAABULRXiE6Leffddbd68WSdOnFBYWJh69uyp2267TXa7XZKUkpKil156STt37lRISIhuvvlm/e1vfwtw1QAAACgun46c9unTx+tmJ9M09cQTT+iPP/7wGvftt9+qUaNG/q3wP/sbO3aslixZoqeeekpfffWVVq1a5Vm/YMECud1uvfHGG3r88ce1ZMkSff/9936vAwAAAGXLp3C6YcMGnTlzxvM4NzdXU6dO1e+//+41LjMzU4cOHfJvhZIGDBigiy66SEFBQYqLi1P37t21e/duSVJGRoY+/fRTDRkyRKGhoWrcuLF69eqlDRs2+L0OAAAAlC2fP770XD7M3V9mfvzxR9WrV0+SdPToUUnyPJakRo0alUlIBgAACJRXXnlFtWvXVnh4uPbv3x/ocspMwK85Pd8ncORdV5pn3bp1OnjwoO677z5JZ4+choSEeI0JCwtTenp6gdtLSEjw+lQom82m2NjYkpRe6eS91+e+5/CdYRi8f6VAD/oHfVhy9KB/VMUeNLOzlbn7B+UmJckWGSlX85YyHA6/bT87O1v33nuvPv74Y7Vr107S2ff5p59+0sUXX+y3/VhBwMPp448/rl27dhW4rlq1anrzzTc9jzdv3qx33nlH06dPV2RkpCQpODg4XxBNTU3NF1jzrFixQgsXLvQ8HjZsmMaMGVPal1Gp5L23KBmn0xnoEio8erD06MPSoQdLryr1YNpnnyhl5XKZWZmS3S653TKcLoX3H6jQTl39so8//vhDGRkZatmypV+2J50NvA4/Bmh/8Tmc7tmzR0FBZ4fnHe38+eefvcac+9gXM2fO9Gnc1q1b9cYbb2jatGmKj4/3LK9bt64k6ciRI7rgggskSQcOHFD9+vUL3M6AAQPUvXt3z2ObzabExMRi110Z2e12RUZGKikpic+ULqGwsDClpqYGuowKix70D/qw5OhB/yiPHoyOji7T7fsq7bNPlPy/b0l5lzv+Z0pNMyP97HKpWAH12Wef1YIFC3T8+HHFx8dr+vTpat68uS6//HJJUkxMjJo2baqwsDBJUtu2bWUYhp5//nnddddd+uqrrzRu3Djt2rVLtWrV0owZM3TTTTdJOntAzuVy6dixY9q0aZNeffVVDRo0yF9vhd/4HE6HDRuWb9mQIUNkGIbnsWmaXo/9ZevWrVq4cKGmTp2aL3QGBwerc+fOWrJkicaOHavjx49r48aNmjBhQoHbiomJUUxMjOdxQkICP4DO4Xa7eU9KyDRN3js/oAdLhz4sPXqwdKpKD5rZ2UpZufy/wTTfAFMpq5YrpF1Hn0/xN2zYUFu3blXt2rW1fPlyDRkyRPv27dOPP/6ohg0bKiEhwTN5v2EY+vrrrz2n9Y8dO6ZrrrlGr732mm644Qbt2LFD1157rZo1a6ZmzZpJkpYsWaK1a9dq9erVysjIKP2bUAZ8CqebN28u6zqK9NZbbyk1NVUPP/ywZ1nz5s01ZcoUSWc/JGDu3LkaNmyYQkNDNXjwYLVq1SpA1QIAgKogc/cPZ0/lF8HMzFTm7l0KbtXGp23+/e9/9/z9lltu0ZNPPqnt27erbdu2533uW2+9pd69e+vGG2+UJHXo0EH9+/fX8uXLNWnSJEnS9ddfr549e0pSoZdABppP4fSvp8ED4dVXXy1yfXh4uCZOnFhO1QAAAEi5SUlnrzEt6tMx7XblJp0pfP053nzzTT3//PM6ePCgpLMfNPTXG7mLcvDgQa1evVrVqlXzLMvJydE//vEPz+O/zm5kVQG/IQoAAKAiskVGSue7fMHtli0yyqftHTp0SCNGjNCGDRvUuXNn2e12tWnTxufpO+vVq6dBgwZp0aJFhddsK/EsouXGp3Dat29fnzdoGIZWr15d4oIAAAAqAlfzljKcLpkZBU9fKUmGyyVX8xY+bS/vJrK8KS7ffPPNQmc0kqSaNWvq119/9VxzOmTIELVt21Zr167Vtddeq9zcXH3zzTeKjIz0XHNaEfgUn5OSkpScnFzk19GjR7Vu3TqtW7eurGsGAAAIOMPhUHj/gVJhN4MbhsJvHOjzzVDNmzfX+PHj1blzZ9WsWVPffvutOnXqVOj4qVOnasSIEapWrZoWLlyo+Ph4vffee5o9e7Zq1qyp2rVr6+GHH1ZmZtHXxVqNYZbyo56Sk5M1Z84cvfDCC0pNTdXIkSM1Z84cf9VX5ny9jqMqsNvtio6OVmJiYpW4y7IsREREKDk5OdBlVFj0oH/QhyVHD/pHefTgX2feKWvney1pn32ilFXLZWb+ZZ5Tl0vhNxY+z6nD4fDcdQ9vJb7mNDk5WS+++KJmz56t9PR0jRw5Ug899JBq167tz/oAAAAsLbRTV4W066jM3buUm3RGtsgouZq38OsnRFUlxQ6nSUlJmj17tl588UVlZGRo1KhRmjBhgmrVqlUW9QEAAFie4XD4PF0UiuZzOD1z5ownlGZlZenuu+/W+PHjFRcXV5b1AQAAoArxKZxOnjxZc+bMkdvt1j//+U89+OCD5XqtBwAAAKoGn8LptGnTZBiGOnTooN27d+uOO+4odCxTSQEAAKCkfAqn3bp1k/GfaRK4AxQAAABlxadwumXLljIuAwAAwJoiIiICXUKVYv3PsAIAAECV4VM4TUhI0Pfff59v+ffff6+///3vuuSSS3TllVdq7dq1fi8QAAAAVYdPp/Uffvhhff3119q5c6dn2aFDh9S1a1elpaWpVatW2rVrl/r3769NmzapW7duZVYwAABAeSqL+234hKjC+XTk9NNPP9XgwYO9lr3wwgtKSUnRe++9px07dujgwYPq2LGjnn766TIpFAAAAJWfT+H06NGjatGihdeytWvXqnXr1urTp48kKSQkRP/zP/9T4Ol/AAAAwBc+hVPDMDxTSUnS8ePHdeDAAXXv3t1rXN26dZWQkODfCgEAAFBl+BROmzZtqg0bNnger1u3ToZheI6a5jl27JhiY2P9WyEAAACqDJ9uiBo7dqxuv/12JSYmqlatWnr55Zd14YUXqnfv3l7j1q9fr5YtW5ZJoQAAAJVdgwYNNH/+fF1zzTXFfu7IkSO1YsUKxcbGas+ePWVQXfnwKZwOHjxYR44c0dy5c3X69Gm1bdtW8+bNU1DQf59+4sQJrV27VlOnTi2zYgEAAKwo2zT1Q3qmkty5irTb1DLEJcdfLoksa9u2bdPatWt16NChCv+hAT5Pwj9x4kT99ttvGjhwoBYtWqRLLrnEa31cXJy2b9+uL7/80u9FAgAAWNUnyWl64MifWvDnGf2/xGQt+POMHjjypz5JTiu3Gg4cOKAGDRqUKJi63W7l5uaWQVUlU+xPiHrzzTf1559/Frju5MmTWrx4camLAgAAqAg+SU7TW6eSlW6aypWUbUq5ktJNU2+dSi5RQP3mm2/UsmVLRUVFacCAATp9+rQk6auvvlLXrl0VHR2tZs2a6d1335UkvfLKKxo5cqS++uorhYeH67777pMkLVq0SE2bNlV0dLR69+6tvXv3evbRoEEDPfXUU2rTpo3CwsJ04sQJJSUladSoUYqPj1etWrU0ZswYZWRklPo9Kq5ih1PTNL3u3P+rffv2qUaNGqUuCgAAwOqyTVPLE1NkFrLelLQ8MUXZZmEjCrZo0SKtXr1av/32mzIzM3Xvvffq2LFjuuaaa/TAAw8oISFBixYt0ogRI/TTTz/prrvu0vz589WuXTulpKRo9uzZ2rJli+6//34tXrxYx48fV6dOnXT99dcrOzvbs5+3335bK1asUFJSkmJjYzV8+HBlZmZq9+7d+vnnn7Vv3z5Nmzat5G9QCfl0zenLL7+sl19+WdLZaaVuu+02hYSEeI3JyMjQwYMHNXDgQP9XCQAAYDE/pGcq8zzBM9M0tSs9U21Cff80qDFjxqhRo0aSpBkzZqh9+/a65JJL1Lt3b914442SpA4dOqh///5avny5Jk2alG8bb7/9toYNG6aOHTtKkiZNmqR58+Zp+/bt6tKlS779nDhxQmvWrNGpU6c8lwY89thjGjp0qGbMmOFz7f7gUzitU6eO2rZtK0natWuXmjZtmm/KKKfTqWbNmunOO+/0f5UAAAAWk+TOld2QcovIp3ZDOuMu3vWc9erV8/y9fv36ysrK0oEDB7R69WpVq1bNsy4nJ0f/+Mc/CtzG0aNHve74DwoKUnx8vI4ePVrgfg4ePCi3260LLrjAs8w0Tbnd7mLV7g8+hdN+/fqpX79+nseTJk1Sw4YNy6woAAAAq4u02+Q+zxl7tylF2Yt3FeXhw4e9/u5wOBQfH69BgwZp0aJFPm2jbt26OnTo0H/rcLv122+/qW7dup5lNtt/66pXr56CgoJ04sQJOZ3OYtXrb8W+5vSNN94gmAIAgCqvZYhLrvNMF+UyDLUIcRVru/PmzdOBAweUnJysxx57TLfccouGDh2qf//731q7dq1ycnKUlZWl7du366effipwG4MHD9bixYu1Y8cOZWVladq0aYqOjlaHDh0KHF+rVi1dd911uvfee5WYmCjTNHXkyBF98MEHxardH4odTgEAACA5DEMDo8NVWDw1JA2MDi/2fKe33367+vbtq/j4eNntdr344ouKj4/Xe++9p9mzZ6tmzZqqXbu2Hn74YWVmZha4jZ49e+qZZ57Rbbfdpri4OH388cdau3atHA5HoftdvHixgoKC1Lp1a0VFRenqq6/2usO/vBimWcxbyCqZhISEQJdgGXa7XdHR0UpMTAzINSaVQUREhJKTkwNdRoVFD/oHfVhy9KB/lEcPxsTElOn2/+p8r+WT5DQtT0xRpmnKbpw9le/6T3DtGhFa4HMcDoeCg32/Saoq8emaUwAAABSsa0SoOoaHaFd6ps64cxVlt6lFOX9CVGVCOAUAACglh2EUa7ooFI5rTgEAAGAZhFMAAABYBuEUAAAAlsE1pwAAAEWw2+1+3+ZfJ8CHN8IpAABAEUJDC54OCmWD2A4AAADLIJwCAADAMginAAAAsAzCKQAAACyDcAoAAADLIJwCAADAMginAAAAsAzCKQAAACyDcAoAAADLIJwCAADAMginAAAAsAzCKQAAACzDME3TDHQRgZSUlCSXyxXoMizBMAw5nU5lZWWpirdFiQUFBSknJyfQZVRY9KB/0IclRw/6R3n0IL+7K6+gQBcQaFlZWcrKygp0GZZgt9vldDqVmpoqt9sd6HIqpIiICCUnJwe6jAqLHvQP+rDk6EH/KI8eJJxWXpzWBwAAgGUQTgEAAGAZhFMAAABYBuEUAAAAlkE4BQAAgGUQTgEAAGAZhFMAAABYBuEUAAAAlkE4BQAAgGUQTgEAAGAZhFMAAABYBuEUAAAAlkE4BQAAgGUQTgEAAGAZhFMAAABYBuEUAAAAlkE4BQAAgGUQTgEAAGAZhFMAAABYBuEUAAAAlkE4BQAAgGUQTgEAAGAZhFMAAABYBuEUAAAAlkE4BQAAgGUQTgEAAGAZhFMAAABYBuEUAAAAlkE4BQAAgGUQTgEAAGAZhFMAAABYBuEUAAAAlkE4BQAAgGUQTgEAAGAZhFMAAABYRlCgCwBQdZjp6cr6eruy9/4smaaCGl8kZ7uOsoWFB7o0AIBFEE4BlIucX39RymsvS+4cye0+u2zvz8r4YJ3Cho6Uo9klAa4QAGAFnNYHUOZyE08p5dWXpKxMTzCVdPbvOTlKXfSK3H8cC1yBAADLIJwCKHOZn30s5ZpFj/lkczlVAwCwMsIpgDKX9d03Z0/nF8btVvb335ZfQQAAyyKcAih72dnnHWLmFBFeAQBVBuEUQJmz16krGUbRY2rVLqdqAABWRjgFUOZcXXoUPcBmk6vrecYAAKoEwimAMhd0cXM52raXbAX8yLHbFdSshRyt25Z/YQAAy2GeUwBlzjAMhd48WFn1Gihz80fKTTx1dnlklFzde8nVtaeMgoIrAKDKIZwCKBeGzSZXp65yXtFFZtKZs8siIgmlAAAvhFMA5cowDBlR1QJdBgDAoircIYtHH31Uffv2VVZWlmdZSkqKnn76ad1yyy0aNmyY3n///QBWCAAAgJKqUEdON27cWODyBQsWyO1264033tCxY8c0adIkxcfH69JLLy3nCgEAAFAaFebIaVJSkpYvX65hw4Z5Lc/IyNCnn36qIUOGKDQ0VI0bN1avXr20YcOGwBQKAACAEqsw4fT1119X3759FRkZ6bX86NGjkqR69ep5ljVq1EiHDh0q1/oAAABQegE/re92u4tcb7fb9cMPP+jIkSMaO3as/vzzT6/1GRkZCgkJ8VoWFham9PT0AreXkJCghIQEz2ObzabY2NgSVl+52O12rz9RfIZh8P6VAj3oH/RhydGD/kEPojQCHk4ff/xx7dq1q8B11apV02uvvab58+frvvvuk62AKWeCg4PzBdHU1NR8gTXPihUrtHDhQs/jYcOGacyYMaV4BZXPuUenUTxOpzPQJVR49GDp0YelQw+WHj2Ikgp4OJ05c2aR648fP66jR49qypQpkiTTNCVJw4cP17hx43TJJZdIko4cOaILLrhAknTgwAHVr1+/wO0NGDBA3bt39zy22WxKTEws7cuoFOx2uyIjI5WUlHTeI9ooWFhYmFJTUwNdRoVFD/oHfVhy9KB/lEcPRkdHl+n2ETgBD6fnExMTo1dffdXzOCEhQRMmTNCsWbNUo0YNOZ1Ode7cWUuWLNHYsWN1/Phxbdy4URMmTCh0ezExMV7b4weQN7fbzXtSQqZp8t75AT1YOvRh6dGDpUMPojQsH07tdrtXmMxr9rxgKkmjRo3S3LlzNWzYMIWGhmrw4MFq1apVQOoFAABAyVk+nJ6rZs2aWrNmjdey8PBwTZw4MUAVAQAAwF8qzFRSAAAAqPwIpwAAALCMCndaH1WTaZrK+Xm3Mj/ZrJzDh2TY7Qpq1kLB3XrKXqduoMsDAAB+QjiF5ZmmqfTV7yjr04/zFsiUlP3NV8r+ertCBw+Ts3XbgNYIAAD8g9P6sLzsb79W1mefSKZ59iuP2y2ZptKWLpb7ZELhGwAAABUG4RSWl7llo5SbW/gAw6asL7aVX0EAAKDMEE5haabbLffRI0UPcucoe8/P5VMQAAAoU4RTAAAAWAbhFJZm2O2yx9crepA9SI6Lm5VPQQAAoEwRTmF5rh69JaOIVjVz5ezYpfwKAgAAZYZwCstztGojZ5fukmF4h1S7XTJsCh08TPbqNQJXIAAA8BvmOYXlGYahkL43ydGsuTI/3iL34YOS3SbHJZfK1aWH7LVqB7pEAADgJ4RTVAiGYcjRpJkcTbi2FACAyozT+gAAALAMwikAAAAsg3AKAAAAyyCcAgAAwDIIpwAAALAMwikAAAAsg3AKAAAAyyCcAgAAwDIIpwAAALAMwikAAAAsg3AKAAAAyyCcAgAAwDIIpwAAALAMwikAAAAsg3AKAAAAyyCcAgAAwDIIpwAAALAMwikAAAAsg3AKAAAAyyCcAgAAwDIIpwAAALAMwikAAAAsg3AKAAAAyyCcAgAAwDIIpwAAALAMwikAAAAsg3AKAAAAyyCcAgAAwDIIpwAAALAMwikAAAAsg3AKAAAAyyCcAgAAwDIIpwAAALAMwikAAAAsg3AKAAAAyzBM0zQDXUQgJSUlyeVyBboMSzAMQ06nU1lZWaribVFiQUFBysnJCXQZFRY96B/0YcnRg/5RHj3I7+7KKyjQBQRaVlaWsrKyAl2GJdjtdjmdTqWmpsrtdge6nAopIiJCycnJgS6jwqIH/YM+LDl60D/KowcJp5UXp/UBAABgGYRTAAAAWAbhFAAAAJZBOAUAAIBlEE4BAABgGYRTAAAAWAbhFAAAAJZBOAUAAIBlEE4BAABgGVX+E6L8yczOUva3O5X94w8yc7IVVK+BnB06yRZVLdClAQAAVAiEUz9x/3FMKQvmyExNk3LPfuRdzr69yvjo3woZeJtc7a8IcIUAAADWx2l9PzAzMpTy8osyU1I9wVSS5M6RTFPpy5cqe9+ewBUIAABQQRBO/SBr51cyM9IlM7fQMRkb15djRQAAABUT4dQPsr//RnK7Cx9gmnL/sldmdlb5FQUAAFABEU79wMzyMXTmFBFgAQAAQDj1B3v8BZLdXuQYIzxCCg4up4oAAAAqJsKpH7iu6Fr0aX27Xa6uPWQYRvkVBQAAUAERTv3AXruOXFdfJxUUPu122evEy9WtZ/kXBgAAUMEwz6mfhFx1rewxscrYsF65x49JkoyQUDk7d1Nwrz4yHM4AVwgAAGB9hFM/cra5XI7WbWWmJEtut4yISBnnuRYVAAAA/0U49TPDMGRERAa6DAAAgAqJa04BAABgGYRTAAAAWAbhFAAAAJZBOAUAAIBlEE4BAABgGYRTAAAAWAbhFAAAAJZBOAUAAIBlEE4BAABgGYRTAAAAWIZhmqYZ6CJgDQkJCVqxYoUGDBigmJiYQJeDKogeRKDRg0DgceQUHgkJCVq4cKESEhICXQqqKHoQgUYPAoFHOAUAAIBlEE4BAABgGfYpU6ZMCXQRsI6QkBBdfvnlCg0NDXQpqKLoQQQaPQgEFjdEAQAAwDI4rQ8AAADLIJwCAADAMginAAAAsIygQBeA8rdu3Tpt2rRJBw8e1BVXXKHx48d71h06dEj/+te/dPDgQdWqVUt33323LrnkkgBWi8ooOztb8+fP13fffafk5GTFxMRo4MCB6tGjhyT6EOVj7ty52rFjh9LT0xUREaE+ffro5ptvlkQPAoHEkdMqqHr16rr55pvVp08fr+U5OTmaPn26rrjiCi1btkwDBgzQjBkzlJKSEqBKUVm53W5Vr15d06dP17Jly3TPPfdo/vz5+vnnn+lDlJu+fftqwYIF+r//+z89+eST2rp1q7Zt20YPAgFGOK2COnXqpI4dOyoyMtJr+Q8//KDMzEz1799fDodDPXv2VM2aNfXZZ58FqFJUVsHBwRo8eLBq1aolm82m5s2bq1mzZvrpp5/oQ5SbevXqyeVyeR4bhqHff/+dHgQCjNP68Dh8+LDq168vm+2//2dp2LChDh8+HMCqUBVkZGTol19+0Q033EAfolwtXrxY69atU2ZmpuLi4tSzZ0999tln9CAQQIRTeKSnpyssLMxrWVhYmNLS0gJUEaoC0zT14osv6qKLLlKbNm20d+9e+hDlZujQobr99tv1yy+/aPv27QoLC+NnIRBgnNaHR0hISL4fvmlpaQoJCQlQRajsTNPUvHnzdPLkSU2YMEGGYdCHKHeGYeiiiy5SUFCQli1bRg8CAUY4hUe9evV06NAh5ebmepYdOHBA9erVC2BVqKxM09T8+fO1f/9+TZkyRcHBwZLoQwRObm6ujh07Rg8CAUY4rYLcbreysrKUm5ur3NxcZWVlKScnRy1btpTD4dCqVauUnZ2trVu36o8//tAVV1wR6JJRCS1YsEB79uzR1KlTvT7DnD5EeUhNTdXmzZuVlvb/27m/kKb6OI7jn9mjoReNZGnzRlGCBsOMRHepIojLbsK6bZIaCjITTQgGtQvpJtqmiEEoBVoXEYJEdBEUEUIJdbHllTQQxPwDFuESbXsuHp7B8h/Pk2s/4v2CA9v3/M53vx8cxodzdraueDyujx8/6tmzZ6qoqOAcBDLMkkgkEpmeBH6viYkJPXr0KKVWV1en7u5uRaNRDQ0NKRqNqrCwUB0dHXI6nRmaKf5US0tLam1tVXGCQgIAAAQuSURBVHZ2tg4dOpSsNzc36+LFi5yHSLv19XUNDAxobm5O8Xhc+fn5qq+v1/nz52WxWDgHgQwinAIAAMAY3NYHAACAMQinAAAAMAbhFAAAAMYgnAIAAMAYhFMAAAAYg3AKAAAAYxBOAQAAYAzCKQAAAIxBOAWQdpOTkxoeHt5WHx8fV1VVlaxWq44cOSKHw6HW1lYtLS0lx9TU1Mhiscjn8207vqamRk1NTcn3L1++lMVi2XGrr69PjpuZmVFLS4scDoeysrJSegAAMuuvTE8AwJ9vcnJSMzMz6uzsTNZu3bql69ev6+rVq/L7/UokEgqHwxofH9fCwoIKCgpSeoRCIfX09Ojo0aP7ft7Y2JhOnjyZUrNarcnXb9680evXr1VdXa1YLPaLqwMAHCTCKYCMGBwclMfj0e3bt5O1xsZG9fX1KR6Pp4ytrq5WOBxWIBDQzZs39+3tdDpVWVm56/6uri55vV5J/1x9BQCYg9v6ANLK4/Ho/v37ikQiyVvsHo9Ha2trstvtOx6TlZX61WSz2dTZ2algMKi1tbVfntPP/QEA5uAbGkBa+Xw+ud1ulZaWanp6WtPT0/L5fDpz5oxGRkZ07949LS4u7tunt7dXm5ubCgQC+4798eOHtra2Urafr8YCAMxEOAWQVmVlZTp27Jhyc3PlcrnkcrlUVlam4eFh5efnq62tTXa7XaWlpfJ6vYpGozv2KSgoUEdHh4LBoL58+bLnZ7pcLmVnZ6dsfr8/DasDABw0wimAjHA6nYpEInr69Km8Xq+sVqtCoZDKy8v14cOHHY/p6+vTxsaGgsHgnr0fPHigd+/epWzt7e3pWAYA4IDxQBSAjMnJyZHb7Zbb7ZYkPX/+XGfPnpXf79eTJ0+2jS8sLNSVK1cUCATU3d29a1+Hw7HnA1EAAHNx5RSAMRoaGnTq1CnNzs7uOubatWuKxWIKhUK/cWYAgN+FcAog7XJycvT9+/eU2ufPn7eNi8Vimp+f1/Hjx3ftZbfb1d7erjt37ujr168HPlcAQGZxWx9A2jkcDo2Ojurhw4c6ceKEbDabqqqqdO7cOTU0NMhut2thYUGDg4NaWVlJ/gfpbvr7+3X37l29f/9eRUVF2/aHw2FtbW2l1A4fPqzTp09LkpaXl/Xq1avk62/fvunx48eSJLfbrby8vINYNgDgfyCcAki7y5cv6+3bt+rq6tLq6qouXbqkGzduaGpqSj09PVpeXpbNZlN5eblevHih2traPfsVFRWpra1NQ0NDO+5vaWnZVisuLk7+E0AkEtGFCxdS9v/7/tOnTyopKfnviwQAHAhLIpFIZHoSAAAAgMRvTgEAAGAQwikAAACMQTgFAACAMQinAAAAMAbhFAAAAMYgnAIAAMAYhFMAAAAYg3AKAAAAYxBOAQAAYAzCKQAAAIxBOAUAAIAx/gaJU6Js5z/cIgAAAABJRU5ErkJggg==\n",
            "text/plain": [
              "<Figure size 640x480 with 1 Axes>"
            ]
          },
          "metadata": {
            "tags": []
          }
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "<ggplot: (285572335)>"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 28
        }
      ]
    },
    {
      "metadata": {
        "colab_type": "text",
        "id": "w7t7-Hjy9ihM"
      },
      "cell_type": "markdown",
      "source": [
        "In both our PCA and our t-SNE graph, we can see a clear distinction between the before-caffeine cluster and the after-caffeine cluster! This indicates that caffeine is changing transcriptome state as the yeast is reacting to the caffeine in its environment on the transcriptome level.\n",
        "\n",
        "Although PCA and t-SNE is often only helpful to indicate that transcripts are being differentially expressed, it is still a crucial first step in data analysis. After this, scientists will often use [heatmaps](https://en.wikipedia.org/wiki/Heat_map), [multidimensional scaling](https://en.wikipedia.org/wiki/Multidimensional_scaling), [statistical tests] (https://pachterlab.github.io/sleuth_walkthroughs/trapnell/analysis.html), or other techniques to continue exploring the secrets hidden within the data. We invite you to continue learning about different crucial methods used by researches and to continue investigating how yeast uses gene regulation to respond to chemical changes in its environment."
      ]
    },
    {
      "metadata": {
        "id": "4-uAEx9pDRn5",
        "colab_type": "text"
      },
      "cell_type": "markdown",
      "source": [
        "# Answers\n",
        "1. Each letter in the fourth line represents the quality of the corresponding nucleotide, so the number of quality scores must match the total number of nucleotides in the sequence.\n",
        "2. We would expect the isoforms.fpkm_tracking file to contain more entries than the genes.fpkm_tracking file, because while humans only have around 25,000 genes, due to alternative splicing, we have a greater number of exons in our transcriptome than number of genes.\n",
        "3. Since we know that ribosomes, the crucial component of the cell that synthesizes proteins, are made out of RNA and proteins, we would expect that a ribosomal subunit to be expressed the most. If you sort the transcripts by tpm value, we find that the highest expressed transcript is VIMSS6581441 with a TPM value of 21603.60. According to microbesOnline, VIMSS6581441 is \"RPL41B Ribosomal protein L47 of the large (60S) ribosomal subunit, identical to Rpl41Ap and has similarity to rat L41 ribosomal protein; comprised of only 25 amino acids; rpl41a rpl41b double null mutant is viable\"."
      ]
    },
    {
      "metadata": {
        "colab_type": "text",
        "id": "QmcIJKSE9ihM"
      },
      "cell_type": "markdown",
      "source": [
        "# Additional Resources\n",
        "\n",
        "- If you are interested in annotation files (`.gff3`), which identify the locations of genes on the genome and the functions of genes, [UCSC Genome Bioinformatics](http://hgdownload.soe.ucsc.edu/downloads.html) is a great starting place to find annotations for your organism.\n",
        "- more technical information on the paired-end sequencing samples used in the original paper can be found [here](https://www.ncbi.nlm.nih.gov/bioproject/PRJNA385168) and [here](https://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?run=SRR5511057). \n",
        "- another common pipline you can explore is: Trimmomatic/STAR/[HTseq](https://htseq.readthedocs.io/en/release_0.11.1/), this provides the expression of transcripts in *counts*, which is great for learning how to do statistical analysis on sequencing data\n",
        "  "
      ]
    },
    {
      "metadata": {
        "colab_type": "text",
        "id": "Q6LoSROz9ihN"
      },
      "cell_type": "markdown",
      "source": [
        "# FAQ\n",
        "\n",
        "\n",
        "#### Q: Why does fastq-dump create two files?\n",
        "> A: Each file for each read. Reads have the same order in both files, they are paired. This [link](https://en.wikipedia.org/wiki/FASTQ_format) explains more details about the format.\n",
        "\n",
        "#### Q: What is Phred quality score?\n",
        "> A: It is a measure of the quality of a single nucleotide read. A phred quality score is encoded by a single ASCII character. [Here](https://en.wikipedia.org/wiki/Phred_quality_score) is a great explanation. In phred +33, this means that 33 is added to the original score, before converting the number to an ascii character.\n",
        "\n",
        "#### Q: After you convert the raw RNA-seq data into each of the major units (counts, FPKM, TPM), what tools do you usually use to analyze and visualize the data before drawing conclusions.\n",
        "> A: About what do we do after we have raw expression data converted into gene expression matrices... well, that's a very good question, because for me expression matrices are just the beginning. Typically an organism like a bacteria has 2,000 genes, so you have 2,000 dimensions per sample. Yeast, 6,000. Humans 25,000 or more, depending on how we define a gene. How to visualize and draw conclusions from two samples with, let's say, 6,000 dimensions, can be challenging. \n",
        "\n",
        "> One of the possible solutions is grouping together genes that have a similar expression pattern. Say, across n samples, m genes coherently change together. That's what we call a bicluster, gene module or regulon. By doing that we tackle two problems: (1) dimensionality reduction, we may go from 6,000 genes to 200 gene modules, and (2) regulatory mechanism inference, i.e., if a set of m genes behaves similarly across n conditions, it is likely that they are modulated by a common upstream regulator, say a transcription factor or another epigenetic mechanism. Those gene modules can be further interrogated to see if they have a common function, like \"cell cycle\" or \"oxidative stress\". That way, by clustering a heatmap of activities of gene modules by samples, we can get closer to understand a biological response. For example, this is heatmap I generated recently, unpublished results! Each column is a single-cell, each row a gene module (each row name refers to the TF regulating n genes, e.g., first row indicates a gene module of 121 genes regulated by transcription factor JUND), color indicates gene module activity. The sequenced cells were therapy resistant melanoma cells. I would say that different cell populations escaped therapy using different transcriptional programs. You can also check out this [site](http://egrin2.systemsbiology.net/) if you feel interested in building gene regulatory networks out of gene expression matrices.\n",
        "\n",
        "#### How do we find the location of a sequence of a spliced transcript in the genome where the sequence is unspliced?\n",
        "> A: What the sequencing machine actually reads are fragments of 75 to 150 base pairs, which means most of the reads are shorter than any exon, so the mapping is quite simple, exons are clearly part of the genome. A less frequent, but clearly possible case would be a read that includes two exons spliced together. Let's say the first 50 base pairs belong to exon 2, and the rest of the read, 100 base pairs belong to exon 3. Although both exons are present in the genome, the joint read sequence does not. Well, those cases are well taken care of. Mapping tools look for those situations and if the read maps to two exons of the same gene, they take that read as successfully mapped. One of the most popular aligners is called [STAR](https://www.ncbi.nlm.nih.gov/pubmed/23104886), which stands for \"Spliced Transcripts Alignment to a Reference\" So it seems that that unfrequent situation was a particularly important problem they solved. You can also check out [this link](https://htseq.readthedocs.io/en/release_0.9.1/count.html) for how reads are considered as valid or not, or ambiguous by HTSeq (another very popular tool) depending on their alignment with respect to features of interest (typically genes, or exons). \n"
      ]
    }
  ]
}